From b0043908542dee74353827ccdb39af40023d67f9 Mon Sep 17 00:00:00 2001
From: RoccoDev <roccodev.business@gmail.com>
Date: Fri, 7 Aug 2020 15:11:00 +0200
Subject: [PATCH] Allow hidden player to be hit by projectiles


diff --git a/src/main/java/net/minecraft/server/EntityArrow.java b/src/main/java/net/minecraft/server/EntityArrow.java
index eac0363c..e9df2f44 100644
--- a/src/main/java/net/minecraft/server/EntityArrow.java
+++ b/src/main/java/net/minecraft/server/EntityArrow.java
@@ -4,6 +4,7 @@ import java.util.List;
 
 // CraftBukkit start
 import org.bukkit.entity.LivingEntity;
+import org.bukkit.entity.Player;
 import org.bukkit.event.entity.EntityCombustByEntityEvent;
 import org.bukkit.event.player.PlayerPickupItemEvent;
 // CraftBukkit end
@@ -222,7 +223,8 @@ public class EntityArrow extends Entity implements IProjectile {
 
             // PaperSpigot start - Allow arrows to fly through vanished players the shooter can't see
             if (movingobjectposition != null && movingobjectposition.entity instanceof EntityPlayer && shooter != null && shooter instanceof EntityPlayer) {
-                if (!((EntityPlayer) shooter).getBukkitEntity().canSee(((EntityPlayer) movingobjectposition.entity).getBukkitEntity())) {
+                Player bukkitEntity = ((EntityPlayer) shooter).getBukkitEntity();
+                if (!bukkitEntity.canSee(((EntityPlayer) movingobjectposition.entity).getBukkitEntity()) && !bukkitEntity.spigot().getCollidesWithEntities()) {
                     movingobjectposition = null;
                 }
             }
diff --git a/src/main/java/net/minecraft/server/EntityFishingHook.java b/src/main/java/net/minecraft/server/EntityFishingHook.java
index df6ce0db..7730337a 100644
--- a/src/main/java/net/minecraft/server/EntityFishingHook.java
+++ b/src/main/java/net/minecraft/server/EntityFishingHook.java
@@ -189,7 +189,8 @@ public class EntityFishingHook extends Entity {
 
             // PaperSpigot start - Allow fishing hooks to fly through vanished players the shooter can't see
             if (movingobjectposition != null && movingobjectposition.entity instanceof EntityPlayer && owner != null && owner instanceof EntityPlayer) {
-                if (!((EntityPlayer) owner).getBukkitEntity().canSee(((EntityPlayer) movingobjectposition.entity).getBukkitEntity())) {
+                Player bukkitEntity = ((EntityPlayer) owner).getBukkitEntity();
+                if (!bukkitEntity.canSee(((EntityPlayer) movingobjectposition.entity).getBukkitEntity()) && !bukkitEntity.spigot().getCollidesWithEntities()) {
                     movingobjectposition = null;
                 }
             }
diff --git a/src/main/java/net/minecraft/server/EntityProjectile.java b/src/main/java/net/minecraft/server/EntityProjectile.java
index e7cca3c6..17e174bc 100644
--- a/src/main/java/net/minecraft/server/EntityProjectile.java
+++ b/src/main/java/net/minecraft/server/EntityProjectile.java
@@ -1,5 +1,9 @@
 package net.minecraft.server;
 
+// KigPaper start
+import org.bukkit.entity.Player;
+// KigPaper end
+
 import java.util.List;
 import java.util.UUID;
 
@@ -149,7 +153,8 @@ public abstract class EntityProjectile extends Entity implements IProjectile {
 
         // PaperSpigot start - Allow projectiles to fly through vanished players the shooter can't see
         if (movingobjectposition != null && movingobjectposition.entity instanceof EntityPlayer && shooter != null && shooter instanceof EntityPlayer) {
-            if (!((EntityPlayer) shooter).getBukkitEntity().canSee(((EntityPlayer) movingobjectposition.entity).getBukkitEntity())) {
+            Player bukkitEntity = ((EntityPlayer) shooter).getBukkitEntity();
+            if (!bukkitEntity.canSee(((EntityPlayer) movingobjectposition.entity).getBukkitEntity()) && !bukkitEntity.spigot().getCollidesWithEntities()) {
                 movingobjectposition = null;
             }
         }
-- 
2.26.2

