From 04db4db55b1aaba30e603aee16dacbb1fb3b71b5 Mon Sep 17 00:00:00 2001
From: RoccoDev <roccodev.business@gmail.com>
Date: Sat, 8 Aug 2020 19:16:21 +0200
Subject: [PATCH] Allow arrows to fly through invis armor stands


diff --git a/src/main/java/net/minecraft/server/EntityArrow.java b/src/main/java/net/minecraft/server/EntityArrow.java
index e9df2f44c..61526c509 100644
--- a/src/main/java/net/minecraft/server/EntityArrow.java
+++ b/src/main/java/net/minecraft/server/EntityArrow.java
@@ -228,6 +228,10 @@ public class EntityArrow extends Entity implements IProjectile {
                     movingobjectposition = null;
                 }
             }
+            // KigPaper start - Allow arrows to fly through invisible armor stands
+            else if (movingobjectposition != null && movingobjectposition.entity instanceof EntityArmorStand && movingobjectposition.entity.isInvisible())
+                movingobjectposition = null;
+            // KigPaper end
             // PaperSpigot end
 
             if (movingobjectposition != null) {
diff --git a/src/main/java/net/minecraft/server/EntityFishingHook.java b/src/main/java/net/minecraft/server/EntityFishingHook.java
index 7730337a6..0c1c99c8f 100644
--- a/src/main/java/net/minecraft/server/EntityFishingHook.java
+++ b/src/main/java/net/minecraft/server/EntityFishingHook.java
@@ -194,6 +194,10 @@ public class EntityFishingHook extends Entity {
                     movingobjectposition = null;
                 }
             }
+            // KigPaper start - Allow arrows to fly through invisible armor stands
+            else if (movingobjectposition != null && movingobjectposition.entity instanceof EntityArmorStand && movingobjectposition.entity.isInvisible())
+                movingobjectposition = null;
+            // KigPaper end
             // PaperSpigot end
 
             if (movingobjectposition != null) {
diff --git a/src/main/java/net/minecraft/server/EntityProjectile.java b/src/main/java/net/minecraft/server/EntityProjectile.java
index 17e174bc8..fa090e9ba 100644
--- a/src/main/java/net/minecraft/server/EntityProjectile.java
+++ b/src/main/java/net/minecraft/server/EntityProjectile.java
@@ -158,6 +158,10 @@ public abstract class EntityProjectile extends Entity implements IProjectile {
                 movingobjectposition = null;
             }
         }
+        // KigPaper start - Allow arrows to fly through invisible armor stands
+        else if (movingobjectposition != null && movingobjectposition.entity instanceof EntityArmorStand && movingobjectposition.entity.isInvisible())
+            movingobjectposition = null;
+        // KigPaper end
         // PaperSpigot end
 
         if (movingobjectposition != null) {
-- 
2.26.2

