From e4d348e409698aee9c51025668b6f6de857529fb Mon Sep 17 00:00:00 2001
From: RoccoDev <roccodev.business@gmail.com>
Date: Tue, 15 Sep 2020 13:54:12 +0200
Subject: [PATCH] Allow disabling of namespaced commands


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index da68aa20a..5abbec2d5 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -1836,5 +1836,9 @@ public final class CraftServer implements Server {
     public String getPermissionMessage() {
         return configuration.getString("settings.permission-message", ChatColor.RED + "I'm sorry, but you do not have permission to perform this command. Please contact the server administrators if you believe that this is in error.");
     }
+    @Override
+    public boolean getAllowNamespaced() {
+        return configuration.getBoolean("settings.namespaced-commands", true);
+    }
     // KigPaper end
 }
-- 
2.26.2

