From 33615c866cf557eb3559161cfaa23a47db32ccf9 Mon Sep 17 00:00:00 2001
From: RoccoDev <roccodev.business@gmail.com>
Date: Mon, 21 Sep 2020 20:24:14 +0200
Subject: [PATCH] Player nickname framework


diff --git a/src/main/java/net/minecraft/server/EntityPlayer.java b/src/main/java/net/minecraft/server/EntityPlayer.java
index 0c49a256..556ab894 100644
--- a/src/main/java/net/minecraft/server/EntityPlayer.java
+++ b/src/main/java/net/minecraft/server/EntityPlayer.java
@@ -66,6 +66,19 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
     public boolean collidesWithEntities = true;
     public int viewDistance; // PaperSpigot - Player view distance API
     private int containerUpdateDelay; // PaperSpigot
+    private String nickname; // KigPaper
+
+    public void setNickname(String nickname) {
+        this.nickname = nickname;
+    }
+
+    public boolean isNicknamed() {
+        return nickname != null;
+    }
+
+    public String getRealName() {
+        return super.getName();
+    }
 
     @Override
     public boolean ad()
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 32eca3bb..eb54b9ec 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -80,6 +80,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
     private double health = 20;
     private boolean scaledHealth = false;
     private double healthScale = 20;
+    private String nickname; // KigPaper
 
     public CraftPlayer(CraftServer server, EntityPlayer entity) {
         super(server, entity);
@@ -258,6 +259,18 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
     }
     // Paper end
 
+    public void setNickname(String nickname) {
+        getHandle().setNickname(nickname);
+    }
+
+    public boolean isNicknamed() {
+        return getHandle().isNicknamed();
+    }
+
+    public String getRealName() {
+        return getHandle().getRealName();
+    }
+
     @Override
     public String getDisplayName() {
         return getHandle().displayName;
-- 
2.26.2

