From b8e64059d9f74ecfb58476da0e66810eb28aafa0 Mon Sep 17 00:00:00 2001
From: RoccoDev <roccodev.business@gmail.com>
Date: Sun, 27 Sep 2020 19:51:13 +0200
Subject: [PATCH] Knockback friction API


diff --git a/src/main/java/dev/rocco/kig/paper/impl/velocity/CraftKnockbackValues.java b/src/main/java/dev/rocco/kig/paper/impl/velocity/CraftKnockbackValues.java
index cf0fd297d..735d1a8f9 100644
--- a/src/main/java/dev/rocco/kig/paper/impl/velocity/CraftKnockbackValues.java
+++ b/src/main/java/dev/rocco/kig/paper/impl/velocity/CraftKnockbackValues.java
@@ -83,4 +83,9 @@ public class CraftKnockbackValues implements KnockbackValues {
     public void resetSprintingVerticalMultiplier() {
         this.sprintingVertical = PaperSpigotConfig.knockbackVerticalSprinting;
     }
+
+    @Override
+    public void setFriction(float friction) {
+        this.friction = friction;
+    }
 }
-- 
2.26.2

