From 70bdad2b34d391fefeff8dfb0d6c898bd337f080 Mon Sep 17 00:00:00 2001
From: RoccoDev <hey@rocco.dev>
Date: Mon, 4 Jan 2021 15:43:19 +0100
Subject: [PATCH] Dynamic server player limit


diff --git a/src/main/java/net/minecraft/server/PlayerList.java b/src/main/java/net/minecraft/server/PlayerList.java
index 3c2d3197f..d566b3177 100644
--- a/src/main/java/net/minecraft/server/PlayerList.java
+++ b/src/main/java/net/minecraft/server/PlayerList.java
@@ -1270,4 +1270,10 @@ public abstract class PlayerList {
     public boolean f(GameProfile gameprofile) {
         return false;
     }
+
+    // KigPaper start
+    public void setMaxPlayers(int maxPlayers) {
+        this.maxPlayers = maxPlayers;
+    }
+    // KigPaper end
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 2cd4e2bdc..c412e31d8 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -1807,10 +1807,13 @@ public final class CraftServer implements Server {
     public SkinCache getSkinCache() {
         return skinCache;
     }
-
     @Override
     public KnockbackValues getKnockbackSettings() {
         return knockbackSettings;
     }
+    @Override
+    public void setMaxPlayers(int count) {
+        getHandle().setMaxPlayers(count);
+    }
     // KigPaper end
 }
-- 
2.28.0

