From a6115acb00587c2e99d5ffaa27d533b661b01985 Mon Sep 17 00:00:00 2001
From: RoccoDev <hey@rocco.dev>
Date: Sun, 17 Jan 2021 17:58:29 +0100
Subject: [PATCH] Add dismount cause to EntityDismountEvent


diff --git a/src/main/java/net/minecraft/server/Entity.java b/src/main/java/net/minecraft/server/Entity.java
index 2da303fee..219e73729 100644
--- a/src/main/java/net/minecraft/server/Entity.java
+++ b/src/main/java/net/minecraft/server/Entity.java
@@ -141,6 +141,14 @@ public abstract class Entity implements ICommandListener {
     public void inactiveTick() { }
     // Spigot end
 
+    // KigPaper start
+    /**
+     * True if the entity is currently evaluating the eject() method from the Bukkit interface, used for
+     * EntityDismountEvent
+     */
+    public boolean ejectingFromBukkit;
+    // KigPaper end
+
     public int getId() {
         return this.id;
     }
@@ -1581,7 +1589,9 @@ public abstract class Entity implements ICommandListener {
                 }
                 // CraftBukkit end
                 // PaperSpigot start - make EntityDismountEvent cancellable
-                EntityDismountEvent dismountEvent = new EntityDismountEvent(this.getBukkitEntity(), this.vehicle.getBukkitEntity()); // Spigot
+                // KigPaper - add cause
+                EntityDismountEvent dismountEvent = new EntityDismountEvent(this.getBukkitEntity(), this.vehicle.getBukkitEntity(),
+                        ejectingFromBukkit ? EntityDismountEvent.DismountCause.BUKKIT : EntityDismountEvent.DismountCause.NATURAL); // Spigot
                 pluginManager.callEvent(dismountEvent);
                 if (dismountEvent.isCancelled()) return;
                 // PaperSpigot end
diff --git a/src/main/java/net/minecraft/server/EntityLiving.java b/src/main/java/net/minecraft/server/EntityLiving.java
index af1a70538..60f901702 100644
--- a/src/main/java/net/minecraft/server/EntityLiving.java
+++ b/src/main/java/net/minecraft/server/EntityLiving.java
@@ -25,8 +25,6 @@ import org.bukkit.event.entity.EntityRegainHealthEvent;
 import org.bukkit.event.vehicle.VehicleExitEvent;
 // CraftBukkit end
 
-import co.aikar.timings.SpigotTimings; // Spigot
-
 // PaperSpigot start
 import org.bukkit.Bukkit;
 import org.spigotmc.event.entity.EntityDismountEvent;
@@ -1724,11 +1722,10 @@ public abstract class EntityLiving extends Entity {
             }
             // CraftBukkit end
             // PaperSpigot start - make dismountEvent cancellable
-            EntityDismountEvent dismountEvent = new EntityDismountEvent(this.getBukkitEntity(), this.vehicle.getBukkitEntity()); // Spigot
-            // KigPaper - only trigger EntityDismountEvent if the dismount was triggered by sneaking
-            if(!(this.bukkitEntity instanceof Player) || ((Player) this.bukkitEntity).isSneaking()) {
-                Bukkit.getPluginManager().callEvent(dismountEvent);
-            }
+            // KigPaper - add cause
+            EntityDismountEvent dismountEvent = new EntityDismountEvent(this.getBukkitEntity(), this.vehicle.getBukkitEntity(),
+                    ejectingFromBukkit ? EntityDismountEvent.DismountCause.BUKKIT : EntityDismountEvent.DismountCause.NATURAL); // Spigot
+            Bukkit.getPluginManager().callEvent(dismountEvent);
             if (dismountEvent.isCancelled()) return;
             // PaperSpigot end
 
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
index 4f9179908..def65a01b 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
@@ -338,8 +338,12 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
         if (getHandle().passenger == null) {
             return false;
         }
-
-        getHandle().passenger.mount(null);
+        // KigPaper start - add cause to EntityDismountEvent
+        Entity passenger = getHandle().passenger;
+        passenger.ejectingFromBukkit = true;
+        passenger.mount(null);
+        passenger.ejectingFromBukkit = false;
+        // KigPaper end
         return true;
     }
 
-- 
2.28.0

