From 40ce6ece85bf5ce2ba4b2ac2149e72faada1b8af Mon Sep 17 00:00:00 2001
From: RoccoDev <hey@rocco.dev>
Date: Wed, 20 Jan 2021 17:28:03 +0100
Subject: [PATCH] Add World getTileEntities


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index 9d1f779e7..0cbe995ed 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -349,6 +349,16 @@ public class CraftWorld implements World {
         // However, this entity is not at the moment backed by a server entity class so it may be left.
         return new CraftItem(world.getServer(), entity);
     }
+
+    @Override
+    public List<BlockState> getTileEntities() {
+        List<BlockState> list = new ArrayList<BlockState>();
+        for (TileEntity entity : world.tileEntityList) {
+            org.bukkit.block.Block block = world.getWorld().getBlockAt(entity.getPosition().getX(), entity.getPosition().getY(), entity.getPosition().getZ());
+            if(block != null) list.add(block.getState());
+        }
+        return list;
+    }
     // KigPaper end
 
     private static void randomLocationWithinBlock(Location loc, double xs, double ys, double zs) {
-- 
2.28.0

