From 9226619dcb0c6feb5ed2101dfd8480073838358a Mon Sep 17 00:00:00 2001
From: RoccoDev <hey@rocco.dev>
Date: Sat, 30 Jan 2021 18:05:31 +0100
Subject: [PATCH] Fix CraftLivingEntity#damage not calling EntityDamageEvent


diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index db915033b..da902708b 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -532,7 +532,8 @@ public class CraftEventFactory {
         } else if (source == DamageSource.FALL) {
             cause = DamageCause.FALL;
         } else if (source == DamageSource.GENERIC) {
-            return new EntityDamageEvent(entity.getBukkitEntity(), null, modifiers, modifierFunctions);
+            //return new EntityDamageEvent(entity.getBukkitEntity(), null, modifiers, modifierFunctions);
+            cause = DamageCause.CUSTOM; // KigPaper
         }
 
         if (cause != null) {
-- 
2.28.0

