From e77f704b3d018d822cc1f03291d97ca5b379edaf Mon Sep 17 00:00:00 2001
From: RoccoDev <hey@rocco.dev>
Date: Wed, 3 Feb 2021 14:07:45 +0100
Subject: [PATCH] Add EntityExtinguishEvent


diff --git a/src/main/java/net/minecraft/server/Entity.java b/src/main/java/net/minecraft/server/Entity.java
index 219e73729..db66228e7 100644
--- a/src/main/java/net/minecraft/server/Entity.java
+++ b/src/main/java/net/minecraft/server/Entity.java
@@ -19,6 +19,7 @@ import org.bukkit.entity.Vehicle;
 import co.aikar.timings.SpigotTimings; // Spigot
 import co.aikar.timings.Timing; // Spigot
 import org.bukkit.event.entity.EntityCombustByEntityEvent;
+import org.bukkit.event.entity.EntityExtinguishEvent;
 import org.bukkit.event.hanging.HangingBreakByEntityEvent;
 import org.bukkit.event.painting.PaintingBreakByEntityEvent;
 import org.bukkit.event.vehicle.VehicleBlockCollisionEvent;
@@ -147,6 +148,10 @@ public abstract class Entity implements ICommandListener {
      * EntityDismountEvent
      */
     public boolean ejectingFromBukkit;
+    /**
+     * Used to detect whether the entity was on fire when fire ticks run out
+     */
+    private boolean wasOnFire;
     // KigPaper end
 
     public int getId() {
@@ -770,8 +775,12 @@ public abstract class Entity implements ICommandListener {
             }
 
             if (flag2 && this.fireTicks > 0) {
+                this.wasOnFire = true; // KigPaper
                 this.makeSound("random.fizz", 0.7F, 1.6F + (this.random.nextFloat() - this.random.nextFloat()) * 0.4F);
                 this.fireTicks = -this.maxFireTicks;
+            } else if(this.wasOnFire && this.fireTicks <= 0) { // KigPaper
+                this.wasOnFire = false;
+                this.world.getServer().getPluginManager().callEvent(new EntityExtinguishEvent(this.getBukkitEntity()));
             }
 
             this.world.methodProfiler.b();
-- 
2.28.0

