From 3bb81871d338258019a82f7cd3b160a7f6f9de80 Mon Sep 17 00:00:00 2001
From: RoccoDev <hey@rocco.dev>
Date: Wed, 3 Feb 2021 15:50:35 +0100
Subject: [PATCH] Fix extinguish event implementation


diff --git a/src/main/java/net/minecraft/server/Entity.java b/src/main/java/net/minecraft/server/Entity.java
index 0880142f3..85a6a6300 100644
--- a/src/main/java/net/minecraft/server/Entity.java
+++ b/src/main/java/net/minecraft/server/Entity.java
@@ -812,13 +812,18 @@ public abstract class Entity implements ICommandListener {
             }
 
             if (flag2 && this.fireTicks > 0) {
-                this.wasOnFire = true; // KigPaper
                 this.makeSound("random.fizz", 0.7F, 1.6F + (this.random.nextFloat() - this.random.nextFloat()) * 0.4F);
                 this.fireTicks = -this.maxFireTicks;
-            } else if(this.wasOnFire && this.fireTicks <= 0) { // KigPaper
+            }
+
+            // KigPaper start
+            if(this.fireTicks > 0) {
+                this.wasOnFire = true;
+            } else if(this.wasOnFire) { // KigPaper
                 this.wasOnFire = false;
                 this.world.getServer().getPluginManager().callEvent(new EntityExtinguishEvent(this.getBukkitEntity()));
             }
+            // KigPaper end
 
             this.world.methodProfiler.b();
         }
-- 
2.28.0

