From e01b1a8d8fbf3c26938a5063a8e35f40c61a35a0 Mon Sep 17 00:00:00 2001
From: RoccoDev <hey@rocco.dev>
Date: Sat, 6 Feb 2021 13:59:56 +0100
Subject: [PATCH] Make forced chunk saving configurable


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index b1ce1da6..86e842e9 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -228,7 +228,8 @@ public class CraftWorld implements World {
             return false;
         }
         // PaperSpigot end
-        if (chunk.mustSave) {   // If chunk had previously been queued to save, must do save to avoid loss of that data
+        if (world.paperSpigotConfig.forceChunkSaveOnUnload /* KigPaper */ && chunk.mustSave) {
+            // If chunk had previously been queued to save, must do save to avoid loss of that data
             save = true;
         }
 
diff --git a/src/main/java/org/github/paperspigot/PaperSpigotWorldConfig.java b/src/main/java/org/github/paperspigot/PaperSpigotWorldConfig.java
index 67e56df0..01f86cae 100644
--- a/src/main/java/org/github/paperspigot/PaperSpigotWorldConfig.java
+++ b/src/main/java/org/github/paperspigot/PaperSpigotWorldConfig.java
@@ -408,4 +408,11 @@ public class PaperSpigotWorldConfig
     {
         disableTeleportationSuffocationCheck = getBoolean("disable-teleportation-suffocation-check", false);
     }
+
+    // KigPaper start
+    public boolean forceChunkSaveOnUnload;
+    private void forceChunkSaveOnUnload() {
+        forceChunkSaveOnUnload = getBoolean("force-chunk-save-unload", true);
+    }
+    // KigPaper end
 }
-- 
2.28.0

