From 64c4c620fd7d740e669e0c31f8e55859e90d8570 Mon Sep 17 00:00:00 2001
From: RoccoDev <hey@rocco.dev>
Date: Mon, 8 Feb 2021 16:49:15 +0100
Subject: [PATCH] Allow disabling cursors in CraftMapRenderer


diff --git a/src/main/java/org/bukkit/craftbukkit/map/CraftMapRenderer.java b/src/main/java/org/bukkit/craftbukkit/map/CraftMapRenderer.java
index ad47bdd4d..74f80c8de 100644
--- a/src/main/java/org/bukkit/craftbukkit/map/CraftMapRenderer.java
+++ b/src/main/java/org/bukkit/craftbukkit/map/CraftMapRenderer.java
@@ -34,16 +34,17 @@ public class CraftMapRenderer extends MapRenderer {
         while (cursors.size() > 0) {
             cursors.removeCursor(cursors.getCursor(0));
         }
-
-        for (UUID key : worldMap.decorations.keySet()) { // Spigot string -> uuid.
-            // If this cursor is for a player check visibility with vanish system
-            Player other = Bukkit.getPlayer(key); // Spigot
-            if (other != null && !player.canSee(other)) {
-                continue;
+        if(getDrawCursors()) { // KigPaper
+            for (UUID key : worldMap.decorations.keySet()) { // Spigot string -> uuid.
+                // If this cursor is for a player check visibility with vanish system
+                Player other = Bukkit.getPlayer(key); // Spigot
+                if (other != null && !player.canSee(other)) {
+                    continue;
+                }
+
+                MapIcon decoration = (MapIcon) worldMap.decorations.get(key);
+                cursors.addCursor(decoration.getX(), decoration.getY(), (byte) (decoration.getRotation() & 15), decoration.getType());
             }
-      
-            MapIcon decoration = (MapIcon) worldMap.decorations.get(key);
-            cursors.addCursor(decoration.getX(), decoration.getY(), (byte) (decoration.getRotation() & 15), decoration.getType());
         }
     }
 
-- 
2.28.0

