From 3b8a28ccf12f404d88d0ad04779623550e5fda74 Mon Sep 17 00:00:00 2001
From: RoccoDev <hey@rocco.dev>
Date: Thu, 11 Feb 2021 17:23:50 +0100
Subject: [PATCH] Add MapView Unload API


diff --git a/src/main/java/net/minecraft/server/PersistentCollection.java b/src/main/java/net/minecraft/server/PersistentCollection.java
index 6459594c7..5aa6e8d8b 100644
--- a/src/main/java/net/minecraft/server/PersistentCollection.java
+++ b/src/main/java/net/minecraft/server/PersistentCollection.java
@@ -187,4 +187,15 @@ public class PersistentCollection {
             return oshort.shortValue();
         }
     }
+
+    // KigPaper start
+    public void removeMap(String mapId, PersistentBase map) {
+        this.c.remove(this.a.remove(mapId));
+        if(map != null) this.c.remove(map); // in case the lookup by ID failed
+        File toDelete = b.getDataFile(mapId);
+        if(toDelete == null || !toDelete.delete()) {
+            MinecraftServer.LOGGER.warn("Couldn't delete file for PersistentBase of ID '" + mapId + "'");
+        }
+    }
+    // KigPaper end
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/map/CraftMapView.java b/src/main/java/org/bukkit/craftbukkit/map/CraftMapView.java
index 1a150d98b..767e8b70c 100644
--- a/src/main/java/org/bukkit/craftbukkit/map/CraftMapView.java
+++ b/src/main/java/org/bukkit/craftbukkit/map/CraftMapView.java
@@ -160,4 +160,16 @@ public final class CraftMapView implements MapView {
         return render;
     }
 
+    // KigPaper start
+    @Override
+    public void unload() {
+        // Remove all renderers
+        while(!renderers.isEmpty()) removeRenderer(renderers.get(0));
+        net.minecraft.server.World worldMain = ((CraftWorld) Bukkit.getServer().getWorlds().get(0)).getHandle();
+        // Remove map from main world's saved maps (including file)
+        worldMain.worldMaps.removeMap(worldMap.id, worldMap);
+        // Clear render cache
+        renderCache.clear();
+    }
+    // KigPaper end
 }
-- 
2.28.0

