From 9383434b5f29adcf27b943bfc5aef11a1f9d4bfa Mon Sep 17 00:00:00 2001
From: RoccoDev <hey@rocco.dev>
Date: Thu, 11 Feb 2021 17:38:57 +0100
Subject: [PATCH] Fix NPE from 0138 and logging from 0139


diff --git a/src/main/java/net/minecraft/server/EntityInsentient.java b/src/main/java/net/minecraft/server/EntityInsentient.java
index b351add9a..3b806b235 100644
--- a/src/main/java/net/minecraft/server/EntityInsentient.java
+++ b/src/main/java/net/minecraft/server/EntityInsentient.java
@@ -86,7 +86,7 @@ public abstract class EntityInsentient extends EntityLiving {
     }
 
     public EntityLiving getGoalTarget() {
-        return this.goalTarget.get(); // KigPaper - WeakReference
+        return this.goalTarget == null ? null : this.goalTarget.get(); // KigPaper - WeakReference
     }
 
     public void setGoalTarget(EntityLiving entityliving) {
diff --git a/src/main/java/net/minecraft/server/PersistentCollection.java b/src/main/java/net/minecraft/server/PersistentCollection.java
index 5aa6e8d8b..75a760f00 100644
--- a/src/main/java/net/minecraft/server/PersistentCollection.java
+++ b/src/main/java/net/minecraft/server/PersistentCollection.java
@@ -193,8 +193,10 @@ public class PersistentCollection {
         this.c.remove(this.a.remove(mapId));
         if(map != null) this.c.remove(map); // in case the lookup by ID failed
         File toDelete = b.getDataFile(mapId);
-        if(toDelete == null || !toDelete.delete()) {
-            MinecraftServer.LOGGER.warn("Couldn't delete file for PersistentBase of ID '" + mapId + "'");
+        if(toDelete != null && toDelete.exists()) {
+            if(!toDelete.delete()) {
+                MinecraftServer.LOGGER.warn("Couldn't delete file for PersistentBase of ID '" + mapId + "'");
+            }
         }
     }
     // KigPaper end
-- 
2.28.0

