From ae0f186682485d8d0492c5fd743a0c000abe66fa Mon Sep 17 00:00:00 2001
From: RoccoDev <hey@rocco.dev>
Date: Sat, 27 Feb 2021 13:29:31 +0100
Subject: [PATCH] Add Actionbar API


diff --git a/src/main/java/net/minecraft/server/NetworkManager.java b/src/main/java/net/minecraft/server/NetworkManager.java
index 05285bd32..de26f34eb 100644
--- a/src/main/java/net/minecraft/server/NetworkManager.java
+++ b/src/main/java/net/minecraft/server/NetworkManager.java
@@ -209,6 +209,7 @@ public class NetworkManager extends SimpleChannelInboundHandler<Packet> {
     }
 
     private void m() {
+        if (this.i.isEmpty()) return; // KigPaper
         if (this.channel != null && this.channel.isOpen()) {
             this.j.readLock().lock();
 
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 7a33726e0..6c3f8ed1f 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -302,6 +302,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         }
     }
 
+    // KigPaper start
     @Override
     public void updateSelfPlayerInfo() {
         if(getHandle().playerConnection == null) return;
@@ -309,6 +310,15 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         getHandle().playerConnection.sendPacket(new PacketPlayOutPlayerInfo(PacketPlayOutPlayerInfo.EnumPlayerInfoAction.ADD_PLAYER, getHandle()));
     }
 
+    @Override
+    public void sendActionbarMessage(BaseComponent... message) {
+        if(getHandle().playerConnection == null) return;
+        PacketPlayOutChat packet = new PacketPlayOutChat(null, (byte) 2);
+        packet.components = message;
+        getHandle().playerConnection.sendPacket(packet);
+    }
+    // KigPaper end
+
     @Override
     public boolean equals(Object obj) {
         if (!(obj instanceof OfflinePlayer)) {
-- 
2.28.0

