From 8b01575f5a0019f69c6e23bf67348a266cee18c2 Mon Sep 17 00:00:00 2001
From: RoccoDev <hey@rocco.dev>
Date: Fri, 5 Mar 2021 16:02:56 +0100
Subject: [PATCH] Fix canFlush from 0147


diff --git a/src/main/java/net/minecraft/server/NetworkManager.java b/src/main/java/net/minecraft/server/NetworkManager.java
index 8d0a7459d..62d6c504a 100644
--- a/src/main/java/net/minecraft/server/NetworkManager.java
+++ b/src/main/java/net/minecraft/server/NetworkManager.java
@@ -86,7 +86,7 @@ public class NetworkManager extends SimpleChannelInboundHandler<Packet> {
     public void disableAutomaticFlush() {
         synchronized (this.flushLock) {
             this.flushPacketsStart = this.packetWrites.get();
-            this.canFlush = true;
+            this.canFlush = false;
         }
     }
 
-- 
2.28.0

