From 5b5868d1b11f6157e2d92b229c1b71d9d9a3847a Mon Sep 17 00:00:00 2001
From: RoccoDev <hey@rocco.dev>
Date: Fri, 5 Mar 2021 16:59:23 +0100
Subject: [PATCH] Add experimental conditional auto-flush for regular
 dispatches


diff --git a/src/main/java/dev/rocco/kig/paper/impl/command/KigPaperCommand.java b/src/main/java/dev/rocco/kig/paper/impl/command/KigPaperCommand.java
new file mode 100644
index 000000000..4ed6e000b
--- /dev/null
+++ b/src/main/java/dev/rocco/kig/paper/impl/command/KigPaperCommand.java
@@ -0,0 +1,24 @@
+package dev.rocco.kig.paper.impl.command;
+
+import net.minecraft.server.NetworkManager;
+import org.bukkit.command.Command;
+import org.bukkit.command.CommandSender;
+
+public class KigPaperCommand extends Command {
+    public KigPaperCommand() {
+        super("kigpaper");
+        this.setPermission("kig.paper.command");
+    }
+
+    @Override
+    public boolean execute(CommandSender sender, String commandLabel, String[] args) {
+        if (!testPermission(sender)) return true;
+        if(args.length > 0) {
+            if("flush".equalsIgnoreCase(args[0])) {
+                NetworkManager.EXPERIMENTAL_FORCE_CONDITIONAL_FLUSH = !NetworkManager.EXPERIMENTAL_FORCE_CONDITIONAL_FLUSH;
+                sender.sendMessage("Experimental force conditional flush: " + NetworkManager.EXPERIMENTAL_FORCE_CONDITIONAL_FLUSH);
+            }
+        }
+        return true;
+    }
+}
diff --git a/src/main/java/net/minecraft/server/NetworkManager.java b/src/main/java/net/minecraft/server/NetworkManager.java
index 62d6c504a..ff6f0d73a 100644
--- a/src/main/java/net/minecraft/server/NetworkManager.java
+++ b/src/main/java/net/minecraft/server/NetworkManager.java
@@ -30,7 +30,7 @@ import org.apache.logging.log4j.Marker;
 import org.apache.logging.log4j.MarkerManager;
 
 public class NetworkManager extends SimpleChannelInboundHandler<Packet> {
-
+    public static boolean EXPERIMENTAL_FORCE_CONDITIONAL_FLUSH = false;
     private static final Logger g = LogManager.getLogger();
     public static final Marker a = MarkerManager.getMarker("NETWORK");
     public static final Marker b = MarkerManager.getMarker("NETWORK_PACKETS", NetworkManager.a);
@@ -168,7 +168,7 @@ public class NetworkManager extends SimpleChannelInboundHandler<Packet> {
     public void handle(Packet packet) {
         if (this.g()) {
             this.m();
-            this.a(packet, (GenericFutureListener[]) null, true /* KigPaper */);
+            this.a(packet, (GenericFutureListener[]) null, !EXPERIMENTAL_FORCE_CONDITIONAL_FLUSH || this.canFlush /* KigPaper */);
         } else {
             this.j.writeLock().lock();
 
diff --git a/src/main/java/org/spigotmc/SpigotConfig.java b/src/main/java/org/spigotmc/SpigotConfig.java
index b9d185d99..e2df64a26 100644
--- a/src/main/java/org/spigotmc/SpigotConfig.java
+++ b/src/main/java/org/spigotmc/SpigotConfig.java
@@ -14,6 +14,8 @@ import java.util.Map;
 import java.util.Set;
 import java.util.concurrent.TimeUnit;
 import java.util.logging.Level;
+
+import dev.rocco.kig.paper.impl.command.KigPaperCommand;
 import gnu.trove.map.hash.TObjectIntHashMap;
 import com.google.common.collect.Lists;
 import net.minecraft.server.AttributeRanged;
@@ -424,4 +426,10 @@ public class SpigotConfig
             Bukkit.getLogger().info( "Debug logging is disabled" );
         }
     }
+
+    // KigPaper start
+    private static void kigPaperCommand() {
+        commands.put("kigpaper", new KigPaperCommand());
+    }
+    // KigPaper end
 }
-- 
2.28.0

