From fe95c6b1310aba87b396e8f2565e63fc772293c8 Mon Sep 17 00:00:00 2001
From: RoccoDev <hey@rocco.dev>
Date: Wed, 10 Mar 2021 20:21:50 +0100
Subject: [PATCH] Add LivingEntity clear last attacked API


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
index 164f19ed..40585f78 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
@@ -463,6 +463,7 @@ public class CraftLivingEntity extends CraftEntity implements LivingEntity {
         return true;
     }
 
+    // KigPaper start
     @Override
     public void setCanArrowsBeStuck(boolean capability) {
         getHandle().setStickArrows(capability);
@@ -474,6 +475,12 @@ public class CraftLivingEntity extends CraftEntity implements LivingEntity {
         return getHandle().getStickArrows();
     }
 
+    @Override
+    public void clearLastAttacked() {
+        getHandle().p(null);
+    }
+    // KigPaper end
+
     @Deprecated
     public int _INVALID_getLastDamage() {
         return NumberConversions.ceil(getLastDamage());
-- 
2.28.0

