From 58e58faac47a861957b4f7a9f746ebfef077f808 Mon Sep 17 00:00:00 2001
From: RoccoDev <hey@rocco.dev>
Date: Sun, 4 Apr 2021 17:58:11 +0200
Subject: [PATCH] Add particle emitter API to World.Spigot#playEffect


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index d6fdbcb6a..555481816 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -1469,7 +1469,8 @@ public class CraftWorld implements World {
     private final Spigot spigot = new Spigot()
     {
         @Override
-        public void playEffect( Location location, Effect effect, int id, int data, float offsetX, float offsetY, float offsetZ, float speed, int particleCount, int radius )
+        // KigPaper - implement method with emitter
+        public void playEffect( Location location, Effect effect, int id, int data, float offsetX, float offsetY, float offsetZ, float speed, int particleCount, int radius, Player emitter )
         {
             Validate.notNull( location, "Location cannot be null" );
             Validate.notNull( effect, "Effect cannot be null" );
@@ -1519,6 +1520,9 @@ public class CraftWorld implements World {
                 {
                     continue;
                 }
+                // KigPaper start - hide particles to players that can't see the emitter
+                if (emitter != null && player != emitter && !player.canSee(emitter)) continue;
+                // KigPaper end
                 distance = (int) player.getLocation().distanceSquared( location );
                 if ( distance <= radius )
                 {
-- 
2.28.0

