From 265e90c14163fc5ddba30c95f82b516b04161468 Mon Sep 17 00:00:00 2001
From: RoccoDev <hey@rocco.dev>
Date: Mon, 12 Apr 2021 16:46:05 +0200
Subject: [PATCH] Allow teleporting vehicles when PlayerMoveEvent is cancelled


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 6c3f8ed1f..d83207cb0 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -562,7 +562,11 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         }
 
         if (entity.passenger != null) {
-            return false;
+            // KigPaper - only cancel if not unknown
+            // if a PlayerMoveEvent is cancelled using setTo and the player has a passenger, the event won't set the user back.
+            if (cause != PlayerTeleportEvent.TeleportCause.UNKNOWN) {
+                return false;
+            }
         }
 
         // From = Players current Location
-- 
2.28.0

