From 089b979b93f1fb6e8a11b9c91408c2b3607b05c2 Mon Sep 17 00:00:00 2001
From: RoccoDev <hey@rocco.dev>
Date: Wed, 14 Apr 2021 14:00:06 +0200
Subject: [PATCH] Add exit code to shutdown API


diff --git a/src/main/java/net/minecraft/server/DedicatedServer.java b/src/main/java/net/minecraft/server/DedicatedServer.java
index 415f235e7..a6232d74c 100644
--- a/src/main/java/net/minecraft/server/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/DedicatedServer.java
@@ -39,6 +39,7 @@ public class DedicatedServer extends MinecraftServer implements IMinecraftServer
     private boolean generateStructures;
     private WorldSettings.EnumGamemode r;
     private boolean s;
+    private int exitStatus; // KigPaper
 
     // CraftBukkit start - Signature changed
     public DedicatedServer(joptsimple.OptionSet options) {
@@ -374,9 +375,15 @@ public class DedicatedServer extends MinecraftServer implements IMinecraftServer
     }
 
     protected void z() {
-        System.exit(0);
+        System.exit(exitStatus); // KigPaper - add exit status
     }
 
+    // KigPaper start
+    public void setExitStatus(int exitStatus) {
+        this.exitStatus = exitStatus;
+    }
+    // KigPaper end
+
     public void B() { // CraftBukkit - fix decompile error
         super.B();
         this.aO();
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index d6d979175..c9630d7c4 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -50,7 +50,6 @@ import org.bukkit.craftbukkit.util.CraftMagicNumbers;
 import org.bukkit.craftbukkit.util.DatFileFilter;
 import org.bukkit.craftbukkit.util.Versioning;
 import org.bukkit.craftbukkit.util.permissions.CraftDefaultPermissions;
-import org.bukkit.entity.Entity;
 import org.bukkit.entity.Player;
 import org.bukkit.event.inventory.InventoryType;
 import org.bukkit.event.player.PlayerChatTabCompleteEvent;
@@ -1277,8 +1276,16 @@ public final class CraftServer implements Server {
 
     @Override
     public void shutdown() {
+        shutdown(0);
+    }
+
+    // KigPaper start
+    @Override
+    public void shutdown(int exitCode) {
+        ((DedicatedServer) console).setExitStatus(exitCode);
         console.safeShutdown();
     }
+    // KigPaper end
 
     @Override
     public int broadcast(String message, String permission) {
-- 
2.28.0

