From a7ca693229c15063256ecabb34fc67894dddaf22 Mon Sep 17 00:00:00 2001
From: RoccoDev <hey@rocco.dev>
Date: Fri, 16 Apr 2021 20:42:20 +0200
Subject: [PATCH] Add option to disable player files


diff --git a/src/main/java/net/minecraft/server/PlayerList.java b/src/main/java/net/minecraft/server/PlayerList.java
index a1842c06e..6df308a86 100644
--- a/src/main/java/net/minecraft/server/PlayerList.java
+++ b/src/main/java/net/minecraft/server/PlayerList.java
@@ -37,6 +37,7 @@ import org.bukkit.event.player.PlayerQuitEvent;
 import org.bukkit.event.player.PlayerRespawnEvent;
 import org.bukkit.event.player.PlayerTeleportEvent.TeleportCause;
 import org.bukkit.util.Vector;
+import org.github.paperspigot.PaperSpigotConfig;
 import org.spigotmc.event.player.PlayerSpawnLocationEvent;
 // CraftBukkit end
 
@@ -296,6 +297,7 @@ public abstract class PlayerList {
     }
 
     protected void savePlayerFile(EntityPlayer entityplayer) {
+        if (!PaperSpigotConfig.savePlayerFiles) return; // KigPaper
         this.playerFileData.save(entityplayer);
         ServerStatisticManager serverstatisticmanager = (ServerStatisticManager) this.o.get(entityplayer.getUniqueID());
 
diff --git a/src/main/java/net/minecraft/server/WorldNBTStorage.java b/src/main/java/net/minecraft/server/WorldNBTStorage.java
index eb195b92e..d03aac0fd 100644
--- a/src/main/java/net/minecraft/server/WorldNBTStorage.java
+++ b/src/main/java/net/minecraft/server/WorldNBTStorage.java
@@ -15,6 +15,7 @@ import org.apache.logging.log4j.Logger;
 import java.util.UUID;
 
 import org.bukkit.craftbukkit.entity.CraftPlayer;
+import org.github.paperspigot.PaperSpigotConfig;
 import org.github.paperspigot.exception.ServerInternalException;
 // CraftBukkit end
 
@@ -201,6 +202,7 @@ public class WorldNBTStorage implements IDataManager, IPlayerFileData {
     }
 
     public NBTTagCompound load(EntityHuman entityhuman) {
+        if (!PaperSpigotConfig.savePlayerFiles) return null; // KigPaper
         NBTTagCompound nbttagcompound = null;
 
         try {
diff --git a/src/main/java/org/github/paperspigot/PaperSpigotConfig.java b/src/main/java/org/github/paperspigot/PaperSpigotConfig.java
index 0ed3fa49d..f628a8bc0 100644
--- a/src/main/java/org/github/paperspigot/PaperSpigotConfig.java
+++ b/src/main/java/org/github/paperspigot/PaperSpigotConfig.java
@@ -224,7 +224,7 @@ public class PaperSpigotConfig
     }
     // KigPaper start
     public static float knockbackHorizontalMultiplier, knockbackVerticalMultiplier, knockbackHorizontalSprinting, knockbackVerticalSprinting, knockbackFriction;
-    public static boolean betterVehicleHitboxes, nettyReadTimeout, forceConditionalAutoFlush;
+    public static boolean betterVehicleHitboxes, nettyReadTimeout, forceConditionalAutoFlush, savePlayerFiles;
     private static void knockbackHorizontalMultiplier() {
         knockbackHorizontalMultiplier = (float) getDouble("knockback.horizontal.normal", 1f);
     }
@@ -245,5 +245,8 @@ public class PaperSpigotConfig
     private static void conditionalAutoFlush() {
         forceConditionalAutoFlush = getBoolean("net-force-conditional-flush", true);
     }
+    private static void savePlayerFiles() {
+        savePlayerFiles = getBoolean("save-player-files", true);
+    }
     // KigPaper end
 }
-- 
2.28.0

