From d5ad2c5338950d152fff9455b7ccdef0125105c1 Mon Sep 17 00:00:00 2001
From: RoccoDev <hey@rocco.dev>
Date: Sat, 15 May 2021 21:14:25 +0200
Subject: [PATCH] Don't make PlayerInfoData static


diff --git a/src/main/java/net/minecraft/server/PacketPlayOutPlayerInfo.java b/src/main/java/net/minecraft/server/PacketPlayOutPlayerInfo.java
index 890c4fc66..95308d8ad 100644
--- a/src/main/java/net/minecraft/server/PacketPlayOutPlayerInfo.java
+++ b/src/main/java/net/minecraft/server/PacketPlayOutPlayerInfo.java
@@ -281,7 +281,7 @@ public class PacketPlayOutPlayerInfo implements Packet<PacketListenerPlayOut> {
         }
     }
 
-    public static class PlayerInfoData { // KigPaper - make static
+    public class PlayerInfoData { // Don't make static - Breaks ProtocolLib
 
         private final int b;
         private final WorldSettings.EnumGamemode c;
-- 
2.28.0

