From 89ce6784173dc8e660e147433b5735d653acebef Mon Sep 17 00:00:00 2001
From: RoccoDev <hey@rocco.dev>
Date: Sun, 23 May 2021 00:25:03 +0200
Subject: [PATCH] Add Player last activity API


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 5f9f92739..0f3aa24aa 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -324,6 +324,12 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         UUID spoofed = getHandle().getNickSpoofedUUID();
         return spoofed == null ? getUniqueId() : spoofed;
     }
+
+    @Override
+    public long getLastActivity() {
+        return getHandle().D();
+    }
+
     // KigPaper end
 
     @Override
-- 
2.31.1

