From 8fcbe2891f4a81615d9e758e1cbabaf9c197328d Mon Sep 17 00:00:00 2001
From: RoccoDev <hey@rocco.dev>
Date: Tue, 1 Jun 2021 13:55:14 +0200
Subject: [PATCH] Player DisconnectReason kick API


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 0f44b3712..2dc4a4ae2 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -331,6 +331,14 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         return getHandle().D();
     }
 
+    @Override
+    public void kickPlayer(DisconnectReason reason, String message) {
+        org.spigotmc.AsyncCatcher.catchOp( "player kick"); // Spigot
+        if (getHandle().playerConnection == null) return;
+
+        getHandle().playerConnection.disconnect(message == null ? "" : message, reason);
+    }
+
     // KigPaper end
 
     @Override
@@ -355,10 +363,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
 
     @Override
     public void kickPlayer(String message) {
-        org.spigotmc.AsyncCatcher.catchOp( "player kick"); // Spigot
-        if (getHandle().playerConnection == null) return;
-
-        getHandle().playerConnection.disconnect(message == null ? "" : message, DisconnectReason.KICK_PLUGIN);
+        kickPlayer(DisconnectReason.KICK_PLUGIN, message);
     }
 
     @Override
-- 
2.31.1

