From 80c6764704785c703c3034af8b031a031b22fd10 Mon Sep 17 00:00:00 2001
From: RoccoDev <hey@rocco.dev>
Date: Fri, 11 Jun 2021 23:02:29 +0200
Subject: [PATCH] Player register plugin channel API


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 2dc4a4ae2..9cc7c3a48 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -339,6 +339,10 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         getHandle().playerConnection.disconnect(message == null ? "" : message, reason);
     }
 
+    @Override
+    public void registerPluginChannel(String channel) {
+        addChannel(channel);
+    }
     // KigPaper end
 
     @Override
-- 
2.31.1

