From a093942ba4b11474cc45794b7a7573c053336784 Mon Sep 17 00:00:00 2001
From: RoccoDev <hey@rocco.dev>
Date: Fri, 18 Jun 2021 22:57:46 +0200
Subject: [PATCH] Add ProjectileCollideEvent


diff --git a/src/main/java/net/minecraft/server/EntityArrow.java b/src/main/java/net/minecraft/server/EntityArrow.java
index 61b83390d..79b9d942c 100644
--- a/src/main/java/net/minecraft/server/EntityArrow.java
+++ b/src/main/java/net/minecraft/server/EntityArrow.java
@@ -1,5 +1,6 @@
 package net.minecraft.server;
 
+import dev.rocco.kig.paper.api.event.ProjectileCollideEvent;
 import dev.rocco.kig.paper.api.particle.ProjectileEffect;
 import org.bukkit.Location;
 import org.bukkit.entity.LivingEntity;
@@ -258,6 +259,13 @@ public class EntityArrow extends Entity implements IProjectile {
             // KigPaper end
             // PaperSpigot end
 
+            // KigPaper start
+            if (movingobjectposition != null && movingobjectposition.entity != null) {
+                ProjectileCollideEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callProjectileCollideEvent(this, movingobjectposition.entity);
+                if (event.isCancelled()) movingobjectposition = null;
+            }
+            // KigPaper end
+
             if (movingobjectposition != null) {
                 org.bukkit.craftbukkit.event.CraftEventFactory.callProjectileHitEvent(this); // CraftBukkit - Call event
                 if (movingobjectposition.entity != null) {
diff --git a/src/main/java/net/minecraft/server/EntityFireball.java b/src/main/java/net/minecraft/server/EntityFireball.java
index 125f74dfc..7a6a0843d 100644
--- a/src/main/java/net/minecraft/server/EntityFireball.java
+++ b/src/main/java/net/minecraft/server/EntityFireball.java
@@ -2,6 +2,7 @@ package net.minecraft.server;
 
 import java.util.List;
 
+import dev.rocco.kig.paper.api.event.ProjectileCollideEvent;
 import org.bukkit.craftbukkit.event.CraftEventFactory; // CraftBukkit
 
 public abstract class EntityFireball extends Entity {
@@ -126,6 +127,13 @@ public abstract class EntityFireball extends Entity {
                 movingobjectposition = new MovingObjectPosition(entity);
             }
 
+            // KigPaper start
+            if (movingobjectposition != null && movingobjectposition.entity != null) {
+                ProjectileCollideEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callProjectileCollideEvent(this, movingobjectposition.entity);
+                if (event.isCancelled()) movingobjectposition = null;
+            }
+            // KigPaper end
+
             if (movingobjectposition != null) {
                 this.a(movingobjectposition);
 
diff --git a/src/main/java/net/minecraft/server/EntityFishingHook.java b/src/main/java/net/minecraft/server/EntityFishingHook.java
index c9c899436..e5d2f80f6 100644
--- a/src/main/java/net/minecraft/server/EntityFishingHook.java
+++ b/src/main/java/net/minecraft/server/EntityFishingHook.java
@@ -1,5 +1,6 @@
 package net.minecraft.server;
 
+import dev.rocco.kig.paper.api.event.ProjectileCollideEvent;
 import dev.rocco.kig.paper.api.particle.ProjectileEffect;
 import org.bukkit.Location;
 import org.bukkit.entity.Fish;
@@ -214,6 +215,13 @@ public class EntityFishingHook extends Entity {
             // KigPaper end
             // PaperSpigot end
 
+            // KigPaper start
+            if (movingobjectposition != null && movingobjectposition.entity != null) {
+                ProjectileCollideEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callProjectileCollideEvent(this, movingobjectposition.entity);
+                if (event.isCancelled()) movingobjectposition = null;
+            }
+            // KigPaper end
+
             if (movingobjectposition != null) {
                 org.bukkit.craftbukkit.event.CraftEventFactory.callProjectileHitEvent(this); // Craftbukkit - Call event
                 if (movingobjectposition.entity != null) {
diff --git a/src/main/java/net/minecraft/server/EntityProjectile.java b/src/main/java/net/minecraft/server/EntityProjectile.java
index eb8336e76..793cd1191 100644
--- a/src/main/java/net/minecraft/server/EntityProjectile.java
+++ b/src/main/java/net/minecraft/server/EntityProjectile.java
@@ -2,6 +2,7 @@ package net.minecraft.server;
 
 // KigPaper start
 
+import dev.rocco.kig.paper.api.event.ProjectileCollideEvent;
 import dev.rocco.kig.paper.api.particle.ProjectileEffect;
 import org.bukkit.Location;
 import org.bukkit.entity.Player;
@@ -177,6 +178,13 @@ public abstract class EntityProjectile extends Entity implements IProjectile {
         // KigPaper end
         // PaperSpigot end
 
+        // KigPaper start
+        if (movingobjectposition != null && movingobjectposition.entity != null) {
+            ProjectileCollideEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callProjectileCollideEvent(this, movingobjectposition.entity);
+            if (event.isCancelled()) movingobjectposition = null;
+        }
+        // KigPaper end
+
         if (movingobjectposition != null) {
             if (movingobjectposition.type == MovingObjectPosition.EnumMovingObjectType.BLOCK && this.world.getType(movingobjectposition.a()).getBlock() == Blocks.PORTAL) {
                 this.d(movingobjectposition.a());
diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index 1f600a17f..ff62ce025 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -9,6 +9,7 @@ import java.util.Map;
 import com.google.common.base.Function;
 import com.google.common.base.Functions;
 
+import dev.rocco.kig.paper.api.event.ProjectileCollideEvent;
 import net.minecraft.server.*;
 
 import org.bukkit.Bukkit;
@@ -764,6 +765,14 @@ public class CraftEventFactory {
         return event;
     }
 
+    // KigPaper start
+    public static ProjectileCollideEvent callProjectileCollideEvent(Entity entity, Entity collided) {
+        ProjectileCollideEvent event = new ProjectileCollideEvent((Projectile) entity.getBukkitEntity(), collided.getBukkitEntity());
+        entity.world.getServer().getPluginManager().callEvent(event);
+        return event;
+    }
+    // KigPaper end
+
     public static ExpBottleEvent callExpBottleEvent(Entity entity, int exp) {
         ThrownExpBottle bottle = (ThrownExpBottle) entity.getBukkitEntity();
         ExpBottleEvent event = new ExpBottleEvent(bottle, exp);
-- 
2.31.1

