From 3c53143ad7df221ad06b84a8186d916c6313a9cd Mon Sep 17 00:00:00 2001
From: RoccoDev <hey@rocco.dev>
Date: Sun, 20 Jun 2021 00:31:39 +0200
Subject: [PATCH] Add compatibility to PlayerConnection


diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index cd3c2c6da..99315de8a 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -161,6 +161,12 @@ public class PlayerConnection implements PacketListenerPlayIn, IUpdatePlayerList
         return this.networkManager;
     }
 
+    // KigPaper - compatibility
+    @Deprecated
+    public void disconnect(String reason) {
+        this.disconnect(reason, DisconnectReason.KICK_PLUGIN);
+    }
+
     public void disconnect(String s, DisconnectReason reason) { // KigPaper - add DisconnectReason
         // KigPaper - if already disconnecting, do nothing
         if(disconnecting) return;
-- 
2.31.1

