From 6aeb09d8cfff9445079a007d04f65f442e4279af Mon Sep 17 00:00:00 2001
From: RoccoDev <hey@rocco.dev>
Date: Sun, 20 Jun 2021 11:43:09 +0200
Subject: [PATCH] Ensure sounds are played to self


diff --git a/src/main/java/net/minecraft/server/PlayerList.java b/src/main/java/net/minecraft/server/PlayerList.java
index cbbd30a45..15926bf55 100644
--- a/src/main/java/net/minecraft/server/PlayerList.java
+++ b/src/main/java/net/minecraft/server/PlayerList.java
@@ -1086,7 +1086,7 @@ public abstract class PlayerList {
             EntityPlayer entityplayer = (EntityPlayer) this.players.get(j);
 
             // CraftBukkit start - Test if player receiving packet can see the source of the packet
-            if (entityhuman != null && entityhuman instanceof EntityPlayer && !entityplayer.getBukkitEntity().canSee(((EntityPlayer) entityhuman).getBukkitEntity())) {
+            if (entityhuman != null && entityplayer != entityhuman && entityhuman instanceof EntityPlayer && !entityplayer.getBukkitEntity().canSee(((EntityPlayer) entityhuman).getBukkitEntity())) {
                continue;
             }
             // CraftBukkit end
@@ -1104,6 +1104,31 @@ public abstract class PlayerList {
 
     }
 
+    // KigPaper start
+    public void sendPacketNearbyIncludingSelf(EntityHuman entityhuman, double d0, double d1, double d2, double d3, int i, Packet packet) {
+        for (int j = 0; j < this.players.size(); ++j) {
+            EntityPlayer entityplayer = (EntityPlayer) this.players.get(j);
+
+            // CraftBukkit start - Test if player receiving packet can see the source of the packet
+            if (entityhuman != null && entityplayer != entityhuman && entityhuman instanceof EntityPlayer && !entityplayer.getBukkitEntity().canSee(((EntityPlayer) entityhuman).getBukkitEntity())) {
+                continue;
+            }
+            // CraftBukkit end
+
+            if (entityplayer.dimension == i) {
+                double d4 = d0 - entityplayer.locX;
+                double d5 = d1 - entityplayer.locY;
+                double d6 = d2 - entityplayer.locZ;
+
+                if (d4 * d4 + d5 * d5 + d6 * d6 < d3 * d3) {
+                    entityplayer.playerConnection.sendPacket(packet);
+                }
+            }
+        }
+
+    }
+    // KigPaper end
+
     public void savePlayers() {
         for (int i = 0; i < this.players.size(); ++i) {
             this.savePlayerFile((EntityPlayer) this.players.get(i));
diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index 9c11ce4e7..e6f55af10 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -948,7 +948,7 @@ public abstract class World implements IBlockAccess {
     public void makeSound(Entity entity, String s, float f, float f1) {
         // KigPaper start - respect visibility for players
         if (entity instanceof EntityHuman) {
-            this.a((EntityHuman) entity, s, f, f1);
+            this.playSoundIncludingSelf((EntityHuman) entity, s, f, f1);
             return;
         }
         // KigPaper end
@@ -966,6 +966,16 @@ public abstract class World implements IBlockAccess {
 
     }
 
+    // KigPaper start
+    public void playSoundIncludingSelf(EntityHuman entityhuman, String s, float f, float f1) {
+        for (IWorldAccess worldAccess : this.u) {
+            if (worldAccess instanceof WorldManager) {
+                ((WorldManager) worldAccess).broadcastPacketIncludingSelf(entityhuman, s, entityhuman.locX, entityhuman.locY, entityhuman.locZ, f, f1);
+            }
+        }
+    }
+    // KigPaper end
+
     public void makeSound(double d0, double d1, double d2, String s, float f, float f1) {
         for (int i = 0; i < this.u.size(); ++i) {
             ((IWorldAccess) this.u.get(i)).a(s, d0, d1, d2, f, f1);
diff --git a/src/main/java/net/minecraft/server/WorldManager.java b/src/main/java/net/minecraft/server/WorldManager.java
index 55a2112d2..2ade8fb5c 100644
--- a/src/main/java/net/minecraft/server/WorldManager.java
+++ b/src/main/java/net/minecraft/server/WorldManager.java
@@ -33,6 +33,13 @@ public class WorldManager implements IWorldAccess {
         this.a.getPlayerList().sendPacketNearby(entityhuman, d0, d1, d2, f > 1.0F ? (double) (16.0F * f) : 16.0D, this.world.dimension, new PacketPlayOutNamedSoundEffect(s, d0, d1, d2, f, f1));
     }
 
+    // KigPaper start
+    public void broadcastPacketIncludingSelf(EntityHuman entityhuman, String s, double d0, double d1, double d2, float f, float f1) {
+        // CraftBukkit - this.world.dimension
+        this.a.getPlayerList().sendPacketNearbyIncludingSelf(entityhuman, d0, d1, d2, f > 1.0F ? (double) (16.0F * f) : 16.0D, this.world.dimension, new PacketPlayOutNamedSoundEffect(s, d0, d1, d2, f, f1));
+    }
+    // KigPaper end
+
     public void a(int i, int j, int k, int l, int i1, int j1) {}
 
     public void a(BlockPosition blockposition) {
-- 
2.31.1

