From 0300166b65c6a16abbc81b3be8a47aba80c3b21f Mon Sep 17 00:00:00 2001
From: RoccoDev <hey@rocco.dev>
Date: Wed, 7 Jul 2021 17:37:36 +0200
Subject: [PATCH] Fix more EnchantmentManager leaks


diff --git a/src/main/java/net/minecraft/server/EnchantmentManager.java b/src/main/java/net/minecraft/server/EnchantmentManager.java
index 9fc38c1d8..63cf45886 100644
--- a/src/main/java/net/minecraft/server/EnchantmentManager.java
+++ b/src/main/java/net/minecraft/server/EnchantmentManager.java
@@ -147,7 +147,9 @@ public class EnchantmentManager {
         } else if (EnchantmentManager.b.a < 0) {
             EnchantmentManager.b.a = 0;
         }
-
+        // KigPaper start
+        EnchantmentManager.b.b = null;
+        // KigPaper end
         return (EnchantmentManager.b.a + 1 >> 1) + EnchantmentManager.a.nextInt((EnchantmentManager.b.a >> 1) + 1);
     }
 
@@ -185,8 +187,8 @@ public class EnchantmentManager {
             a((EnchantmentManager.EnchantmentModifier) EnchantmentManager.e, entityliving.bA());
         }
         // KigPaper start
-        EnchantmentManager.d.a = null;
-        EnchantmentManager.d.b = null;
+        EnchantmentManager.e.a = null;
+        EnchantmentManager.e.b = null;
         // KigPaper end
     }
 
-- 
2.31.1

