From 4d5ab8ccad5e391480774c370412ea677c17cf64 Mon Sep 17 00:00:00 2001
From: RoccoDev <hey@rocco.dev>
Date: Mon, 12 Jul 2021 13:20:21 +0200
Subject: [PATCH] Add chunk conditional block update


diff --git a/src/main/java/net/minecraft/server/Chunk.java b/src/main/java/net/minecraft/server/Chunk.java
index fabf96a40..9f8f7fb56 100644
--- a/src/main/java/net/minecraft/server/Chunk.java
+++ b/src/main/java/net/minecraft/server/Chunk.java
@@ -19,7 +19,6 @@ import org.apache.logging.log4j.Logger;
 import com.google.common.collect.Lists; // CraftBukkit
 import org.bukkit.Bukkit; // CraftBukkit
 import org.bukkit.craftbukkit.util.CraftMagicNumbers;
-import org.github.paperspigot.event.ServerExceptionEvent;
 import org.github.paperspigot.exception.ServerInternalException;
 
 public class Chunk {
@@ -568,8 +567,14 @@ public class Chunk {
         return this.g(blockposition.getX() & 15, blockposition.getY(), blockposition.getZ() & 15);
     }
 
-    // KigPaper - add updateLight param
-    public IBlockData a(BlockPosition blockposition, IBlockData iblockdata, boolean updateLight) {
+    // KigPaper start - add backwards compat & add updateLight param
+    public IBlockData a(BlockPosition blockPosition, IBlockData blockData, boolean updateLight) {
+        return setBlock(blockPosition, blockData, updateLight, true);
+    }
+    // KigPaper end
+
+    // KigPaper - add updateLight param and conditional onPlace
+    public IBlockData setBlock(BlockPosition blockposition, IBlockData iblockdata, boolean updateLight, boolean notifyBlock) {
         int i = blockposition.getX() & 15;
         int j = blockposition.getY();
         int k = blockposition.getZ() & 15;
@@ -640,7 +645,7 @@ public class Chunk {
                 }
 
                 // CraftBukkit - Don't place while processing the BlockPlaceEvent, unless it's a BlockContainer. Prevents blocks such as TNT from activating when cancelled.
-                if (!this.world.isClientSide && block1 != block  && (!this.world.captureBlockStates || block instanceof BlockContainer)) {
+                if (!this.world.isClientSide && notifyBlock && block1 != block  && (!this.world.captureBlockStates || block instanceof BlockContainer)) { // KigPaper - add notifyBlock check
                     block.onPlace(this.world, blockposition, iblockdata);
                 }
 
-- 
2.31.1

