From 9ba61ba405cdbd049e3710c08183d21968d2ff28 Mon Sep 17 00:00:00 2001
From: RoccoDev <hey@rocco.dev>
Date: Fri, 6 Aug 2021 14:55:13 +0200
Subject: [PATCH] Add fake textures to skin cache


diff --git a/src/main/java/net/minecraft/server/EntityPlayer.java b/src/main/java/net/minecraft/server/EntityPlayer.java
index 58e0d1a9..fbcd389c 100644
--- a/src/main/java/net/minecraft/server/EntityPlayer.java
+++ b/src/main/java/net/minecraft/server/EntityPlayer.java
@@ -4,6 +4,7 @@ import com.google.common.base.Charsets;
 import com.google.common.collect.Lists;
 import com.google.common.collect.Sets;
 import com.mojang.authlib.GameProfile;
+import com.mojang.authlib.properties.Property;
 import io.netty.buffer.Unpooled;
 
 import java.util.*;
@@ -70,8 +71,22 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
     private UUID nickSpoofedUUID;
 
     public void setNickname(String nickname) {
+        String oldName = this.nickname;
         this.nickname = nickname;
         this.nickSpoofedUUID = nickname == null ? null : UUID.nameUUIDFromBytes(("KIGNick:" + nickname).getBytes(Charsets.UTF_8));
+
+        if (nickname != null) {
+            // Update skin cache with player's spoofed skin (necessary for skulls to display the nicked player's skin)
+            Collection<Property> fakeTextures = getProfile().getProperties().get("kig-textures");
+            if (!fakeTextures.isEmpty()) {
+                GameProfile forSkinCache = new GameProfile(this.nickSpoofedUUID, nickname);
+                Property oldSkin = fakeTextures.iterator().next();
+                forSkinCache.getProperties().put("textures", new Property("textures", oldSkin.getValue(), oldSkin.getSignature()));
+                TileEntitySkull.skinCache.asMap().put(nickname, forSkinCache);
+            }
+        } else if (oldName != null) {
+            TileEntitySkull.skinCache.invalidate(oldName);
+        }
     }
 
     public boolean isNicknamed() {
-- 
2.31.1

