From 585fe6c65baa9c59c1225c957e934eeda1b046e7 Mon Sep 17 00:00:00 2001
From: RoccoDev <hey@rocco.dev>
Date: Fri, 6 Aug 2021 15:08:11 +0200
Subject: [PATCH] Cache User Authenticator Threads


diff --git a/src/main/java/net/minecraft/server/LoginListener.java b/src/main/java/net/minecraft/server/LoginListener.java
index 20500ce11..1b6cd6b1b 100644
--- a/src/main/java/net/minecraft/server/LoginListener.java
+++ b/src/main/java/net/minecraft/server/LoginListener.java
@@ -82,6 +82,12 @@ public class LoginListener implements PacketLoginInListener, IUpdatePlayerListBo
 
     }
 
+    // KigPaper start - SportPaper-0107
+    private static final java.util.concurrent.ExecutorService authenticatorPool = java.util.concurrent.Executors.newCachedThreadPool(
+            r -> new Thread(r, "User Authenticator #" + b.incrementAndGet())
+    );
+    // KigPaper end
+
     // Spigot start
     public void initUUID()
     {
@@ -166,8 +172,8 @@ public class LoginListener implements PacketLoginInListener, IUpdatePlayerListBo
         } else {
             // Spigot start
             initUUID();
-            new Thread(new Runnable() {
-
+            //new Thread(new Runnable() { // KigPaper
+            authenticatorPool.execute(new Runnable() {
                 @Override
                 public void run() {
                     try{
@@ -177,7 +183,7 @@ public class LoginListener implements PacketLoginInListener, IUpdatePlayerListBo
                         server.server.getLogger().log(java.util.logging.Level.WARNING, "Exception verifying " + i.getName(), ex);
                     }
                 }
-            }).start();
+            });
             // Spigot end
         }
 
@@ -193,7 +199,8 @@ public class LoginListener implements PacketLoginInListener, IUpdatePlayerListBo
             this.loginKey = packetlogininencryptionbegin.a(privatekey);
             this.g = LoginListener.EnumProtocolState.AUTHENTICATING;
             this.networkManager.a(this.loginKey);
-            (new Thread("User Authenticator #" + LoginListener.b.incrementAndGet()) {
+            // (new Thread("User Authenticator #" + LoginListener.b.incrementAndGet()) { // KigPaper - SportPaper-0107
+            authenticatorPool.execute(new Runnable() {
                 public void run() {
                     GameProfile gameprofile = LoginListener.this.i;
 
@@ -233,7 +240,7 @@ public class LoginListener implements PacketLoginInListener, IUpdatePlayerListBo
                     }
 
                 }
-            }).start();
+            });
         }
     }
 
-- 
2.31.1

