From 75b09ca1a7cd29b671a0a3ca1826ac7353192d92 Mon Sep 17 00:00:00 2001
From: RoccoDev <hey@rocco.dev>
Date: Fri, 6 Aug 2021 15:15:20 +0200
Subject: [PATCH] Only save user cache if saving player files is on


diff --git a/src/main/java/net/minecraft/server/PlayerList.java b/src/main/java/net/minecraft/server/PlayerList.java
index 15926bf55..62eeddf6e 100644
--- a/src/main/java/net/minecraft/server/PlayerList.java
+++ b/src/main/java/net/minecraft/server/PlayerList.java
@@ -89,11 +89,12 @@ public abstract class PlayerList {
 
     public void a(NetworkManager networkmanager, EntityPlayer entityplayer) {
         GameProfile gameprofile = entityplayer.getProfile();
+        boolean loadUserCache = PaperSpigotConfig.savePlayerFiles;
         UserCache usercache = this.server.getUserCache();
-        GameProfile gameprofile1 = usercache.a(gameprofile.getId());
+        GameProfile gameprofile1 = loadUserCache ? usercache.a(gameprofile.getId()) : null; // KigPaper - only load if savePlayerFiles is enabled
         String s = gameprofile1 == null ? gameprofile.getName() : gameprofile1.getName();
 
-        usercache.a(gameprofile);
+        if (loadUserCache) usercache.a(gameprofile); // KigPaper - only save if savePlayerFiles is enabled
         NBTTagCompound nbttagcompound = this.a(entityplayer);
         // CraftBukkit start - Better rename detection
         if (nbttagcompound != null && nbttagcompound.hasKey("bukkit")) {
-- 
2.31.1

