From 70570df101813ebb91fe9bb16157b30e7ceefda1 Mon Sep 17 00:00:00 2001
From: RoccoDev <hey@rocco.dev>
Date: Thu, 19 Aug 2021 23:26:38 +0200
Subject: [PATCH] Only change min y for bounding box water check


diff --git a/src/main/java/net/minecraft/server/Entity.java b/src/main/java/net/minecraft/server/Entity.java
index 36d67aecf..0c2b7d5bb 100644
--- a/src/main/java/net/minecraft/server/Entity.java
+++ b/src/main/java/net/minecraft/server/Entity.java
@@ -955,9 +955,14 @@ public abstract class Entity implements ICommandListener {
     }
 
     public boolean W() {
-        if (this.world.a(this.getBoundingBox().grow(0.0D, -0.4000000059604645D +
-                (PaperSpigotConfig.accurateBlockCollisions ? -this.motY : 0.0D), 0.0D) // KigPaper - account for falling speed
-                .shrink(0.001D, 0.001D, 0.001D), Material.WATER, this)) {
+        AxisAlignedBB boundingBox = this.getBoundingBox();
+        // KigPaper start
+        if (PaperSpigotConfig.accurateBlockCollisions) {
+            // Account for vertical speed in minY
+            boundingBox = new AxisAlignedBB(boundingBox.a, boundingBox.b + this.motY, boundingBox.c, boundingBox.d, boundingBox.e, boundingBox.f);
+        }
+        // KigPaper end
+        if (this.world.a(boundingBox.grow(0.0D, -0.4000000059604645D, 0.0D).shrink(0.001D, 0.001D, 0.001D), Material.WATER, this)) {
             if (!this.inWater && !this.justCreated) {
                 this.X();
             }
-- 
2.31.1

