From 2b2b98a9a5eb653094d7f1dc58070316cab71623 Mon Sep 17 00:00:00 2001
From: Archer <archer@beezig.eu>
Date: Mon, 13 Sep 2021 21:27:25 +0200
Subject: [PATCH] accurate-collisions: Switch to Deque


diff --git a/src/main/java/net/minecraft/server/Entity.java b/src/main/java/net/minecraft/server/Entity.java
index 941d79d11..1e1536b07 100644
--- a/src/main/java/net/minecraft/server/Entity.java
+++ b/src/main/java/net/minecraft/server/Entity.java
@@ -850,7 +850,7 @@ public abstract class Entity implements ICommandListener {
         // KigPaper - Start
         double motYAccount = PaperSpigotConfig.accurateBlockCollisions ? this.motY : 0.0D; // KigPaper - account for falling speed
         boolean shouldCollideWithWeb = true, didCollide = false;
-        Stack<Pair<IBlockData, BlockPosition>> websQueue = new Stack<>();
+        Deque<Pair<IBlockData, BlockPosition>> websQueue = new ArrayDeque<>();
         // KigPaper - End
         BlockPosition blockposition = new BlockPosition(this.getBoundingBox().a + 0.001D, this.getBoundingBox().b + 0.001D + motYAccount, this.getBoundingBox().c + 0.001D);
         BlockPosition blockposition1 = new BlockPosition(this.getBoundingBox().d - 0.001D, this.getBoundingBox().e - 0.001D, this.getBoundingBox().f - 0.001D);
-- 
2.33.0

