From ea69b5bc469e987b1f8c3796d73a5a2b26621832 Mon Sep 17 00:00:00 2001
From: Archer <archer@beezig.eu>
Date: Mon, 13 Sep 2021 21:48:10 +0200
Subject: [PATCH] accurate-collisions: Only apply for players


diff --git a/src/main/java/net/minecraft/server/Entity.java b/src/main/java/net/minecraft/server/Entity.java
index 18b895277..22f984a75 100644
--- a/src/main/java/net/minecraft/server/Entity.java
+++ b/src/main/java/net/minecraft/server/Entity.java
@@ -11,10 +11,7 @@ import org.bukkit.Location;
 import org.bukkit.Server;
 import org.bukkit.TravelAgent;
 import org.bukkit.block.BlockFace;
-import org.bukkit.entity.Hanging;
-import org.bukkit.entity.LivingEntity;
-import org.bukkit.entity.Painting;
-import org.bukkit.entity.Vehicle;
+import org.bukkit.entity.*;
 import co.aikar.timings.SpigotTimings; // Spigot
 import co.aikar.timings.Timing; // Spigot
 import org.bukkit.event.entity.*;
@@ -848,7 +845,7 @@ public abstract class Entity implements ICommandListener {
 
     protected void checkBlockCollisions() {
         // KigPaper - Start
-        double motYAccount = PaperSpigotConfig.accurateBlockCollisions ? this.motY : 0.0D; // KigPaper - account for falling speed
+        double motYAccount = this instanceof EntityPlayer && PaperSpigotConfig.accurateBlockCollisions ? this.motY : 0.0D; // KigPaper - account for falling speed
         boolean shouldCollideWithWeb = true, didCollide = false;
         Deque<Pair<IBlockData, BlockPosition>> websQueue = new ArrayDeque<>();
         // KigPaper - End
@@ -866,11 +863,15 @@ public abstract class Entity implements ICommandListener {
 
                         try {
                             // KigPaper - Start
-                            if (shouldCollideWithWeb && iblockdata.getBlock().getMaterial() == Material.WEB) {
-                                websQueue.push(Pair.of(iblockdata, blockposition2));
-                            } else {
+                            if (!PaperSpigotConfig.accurateBlockCollisions || !(this instanceof EntityPlayer)) {
                                 iblockdata.getBlock().a(this.world, blockposition2, iblockdata, this);
-                                if (iblockdata.getBlock().getMaterial().isSolid()) didCollide = true;
+                            } else {
+                                if (shouldCollideWithWeb && iblockdata.getBlock().getMaterial() == Material.WEB) {
+                                    websQueue.push(Pair.of(iblockdata, blockposition2));
+                                } else {
+                                    iblockdata.getBlock().a(this.world, blockposition2, iblockdata, this);
+                                    if (iblockdata.getBlock().getMaterial().isSolid()) didCollide = true;
+                                }
                             }
                             // KigPaper - End
                         } catch (Throwable throwable) {
@@ -883,10 +884,12 @@ public abstract class Entity implements ICommandListener {
                     }
                 }
                 // KigPaper - Start
-                if (didCollide) shouldCollideWithWeb = false;
-                if (shouldCollideWithWeb && !websQueue.isEmpty()) {
-                    for (Pair<IBlockData, BlockPosition> p : websQueue) {
-                        p.getLeft().getBlock().a(this.world, p.getRight(), p.getLeft(), this);
+                if (PaperSpigotConfig.accurateBlockCollisions && this instanceof EntityPlayer) {
+                    if (didCollide) shouldCollideWithWeb = false;
+                    if (shouldCollideWithWeb && !websQueue.isEmpty()) {
+                        for (Pair<IBlockData, BlockPosition> p : websQueue) {
+                            p.getLeft().getBlock().a(this.world, p.getRight(), p.getLeft(), this);
+                        }
                     }
                 }
                 // KigPaper - End
@@ -976,7 +979,7 @@ public abstract class Entity implements ICommandListener {
     public boolean W() {
         AxisAlignedBB boundingBox = this.getBoundingBox();
         // KigPaper start
-        if (PaperSpigotConfig.accurateBlockCollisions) {
+        if (PaperSpigotConfig.accurateBlockCollisions && this instanceof EntityPlayer) {
             // Account for vertical speed in minY
             AxisAlignedBB tmp = new AxisAlignedBB(boundingBox.a, boundingBox.b + this.motY / 2, boundingBox.c, boundingBox.d, boundingBox.e, boundingBox.f);
 
-- 
2.33.0

