From 4a0173c5fb541e9e63a206b6bd4c81f5a75ff540 Mon Sep 17 00:00:00 2001
From: RoccoDev <hey@rocco.dev>
Date: Fri, 10 Dec 2021 00:21:40 +0100
Subject: [PATCH] Disable JNDI lookups


diff --git a/src/main/java/org/apache/logging/log4j/core/lookup/JndiLookup.java b/src/main/java/org/apache/logging/log4j/core/lookup/JndiLookup.java
new file mode 100644
index 000000000..e76be8d59
--- /dev/null
+++ b/src/main/java/org/apache/logging/log4j/core/lookup/JndiLookup.java
@@ -0,0 +1,38 @@
+//
+// Source code recreated from a .class file by IntelliJ IDEA
+// (powered by FernFlower decompiler)
+//
+
+package org.apache.logging.log4j.core.lookup;
+
+import javax.naming.InitialContext;
+import javax.naming.NamingException;
+import org.apache.logging.log4j.core.LogEvent;
+import org.apache.logging.log4j.core.config.plugins.Plugin;
+
+@Plugin(
+    name = "jndi",
+    category = "Lookup"
+)
+public class JndiLookup implements StrLookup {
+    static final String CONTAINER_JNDI_RESOURCE_PATH_PREFIX = "java:comp/env/";
+
+    public JndiLookup() {
+    }
+
+    public String lookup(String var1) {
+        return null;
+    }
+
+    public String lookup(LogEvent var1, String var2) {
+        return null;
+    }
+
+    private String convertJndiName(String var1) {
+        if (!var1.startsWith("java:comp/env/") && var1.indexOf(58) == -1) {
+            var1 = "java:comp/env/" + var1;
+        }
+
+        return var1;
+    }
+}
-- 
2.33.1

