From 570c0063affaba796bc3bcb034ae7aba02542dc7 Mon Sep 17 00:00:00 2001
From: RoccoDev <roccodev.business@gmail.com>
Date: Tue, 15 Sep 2020 13:50:38 +0200
Subject: [PATCH] Allow disabling of namespaced commands


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 9b47a61b..9884ac6e 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -1171,5 +1171,8 @@ public final class Bukkit {
     public static String getPermissionMessage() {
         return server.getPermissionMessage();
     }
+    public static boolean getAllowNamespaced() {
+        return server.getAllowNamespaced();
+    }
     // KigPaper end
 }
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 29b3d88f..0691f96a 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1013,5 +1013,10 @@ public interface Server extends PluginMessageRecipient {
      * @return the server's no permission message
      */
     String getPermissionMessage();
+    /**
+     * @return false if the server should check for the 'kig.namespaced' permission when tab-completing or executing
+     * namespaced commands (prefix:command).
+     */
+    boolean getAllowNamespaced();
     // KigPaper end
 }
diff --git a/src/main/java/org/bukkit/command/SimpleCommandMap.java b/src/main/java/org/bukkit/command/SimpleCommandMap.java
index f35bc092..dd0f5832 100644
--- a/src/main/java/org/bukkit/command/SimpleCommandMap.java
+++ b/src/main/java/org/bukkit/command/SimpleCommandMap.java
@@ -12,6 +12,7 @@ import java.util.Map;
 import java.util.regex.Pattern;
 
 import org.apache.commons.lang.Validate;
+import org.bukkit.Bukkit;
 import org.bukkit.Location;
 import org.bukkit.Server;
 import org.bukkit.command.defaults.*;
@@ -134,6 +135,9 @@ public class SimpleCommandMap implements CommandMap {
         }
 
         String sentCommandLabel = args[0].toLowerCase();
+        // KigPaper start - ignore namespaced commands if enabled and sender is unauthorized
+        if(!Bukkit.getAllowNamespaced() && sentCommandLabel.contains(":") && !sender.hasPermission("kig.namespaced")) return false;
+        // KigPaper end
         Command target = getCommand(sentCommandLabel);
 
         if (target == null) {
@@ -193,6 +197,9 @@ public class SimpleCommandMap implements CommandMap {
             final String prefix = (sender instanceof Player ? "/" : "");
 
             for (Map.Entry<String, Command> commandEntry : knownCommands.entrySet()) {
+                // KigPaper start
+                if(!Bukkit.getAllowNamespaced() && commandEntry.getKey().contains(":") && !sender.hasPermission("kig.namespaced")) continue;
+                // KigPaper end
                 Command command = commandEntry.getValue();
 
                 if (!command.testPermissionSilent(sender)) {
-- 
2.26.2

