From 98f33eb81105613955e1a8783429ae44b9dad698 Mon Sep 17 00:00:00 2001
From: RoccoDev <roccodev.business@gmail.com>
Date: Mon, 21 Sep 2020 16:10:29 +0200
Subject: [PATCH] Item Projectile effects


diff --git a/src/main/java/dev/rocco/kig/paper/api/particle/ItemProjectileEffect.java b/src/main/java/dev/rocco/kig/paper/api/particle/ItemProjectileEffect.java
new file mode 100644
index 00000000..bdc5b525
--- /dev/null
+++ b/src/main/java/dev/rocco/kig/paper/api/particle/ItemProjectileEffect.java
@@ -0,0 +1,24 @@
+package dev.rocco.kig.paper.api.particle;
+
+import org.bukkit.Location;
+import org.bukkit.entity.Item;
+import org.bukkit.inventory.ItemStack;
+import org.bukkit.util.Vector;
+
+public class ItemProjectileEffect extends ProjectileEffect {
+    private ItemStack stack;
+
+    public ItemProjectileEffect(ItemStack stack) {
+        super(1);
+        this.stack = stack;
+    }
+
+    @Override
+    public void spawn(Location loc) {
+        Item item = loc.getWorld().dropImmovableItem(loc, stack);
+        item.setDespawnRate(5);
+        item.noPickup();
+        item.setAffectedByGravity(false);
+        item.setVelocity(new Vector(0, 0, 0));
+    }
+}
diff --git a/src/main/java/org/bukkit/World.java b/src/main/java/org/bukkit/World.java
index 325d65a3..a8137e88 100644
--- a/src/main/java/org/bukkit/World.java
+++ b/src/main/java/org/bukkit/World.java
@@ -1192,6 +1192,15 @@ public interface World extends PluginMessageRecipient, Metadatable {
      */
     public boolean isGameRule(String rule);
 
+    /**
+     * Drops an item at the specified {@link Location}, setting its velocity to 0.
+     *
+     * @param location Location to drop the item
+     * @param item ItemStack to drop
+     * @return ItemDrop entity created as a result of this method
+     */
+    Item dropImmovableItem(Location location, ItemStack item);
+
     // Spigot start
     public class Spigot
     {
diff --git a/src/main/java/org/bukkit/entity/Item.java b/src/main/java/org/bukkit/entity/Item.java
index 90260b7e..6240619c 100644
--- a/src/main/java/org/bukkit/entity/Item.java
+++ b/src/main/java/org/bukkit/entity/Item.java
@@ -34,4 +34,38 @@ public interface Item extends Entity {
      * @param delay New delay
      */
     public void setPickupDelay(int delay);
+
+    /**
+     * @return the item's age
+     */
+    int getAge();
+
+    /**
+     * @return the item's despawn rate. If no custom value is set, this returns the rate according to the Spigot config.
+     */
+    int getDespawnRate();
+
+    /**
+     * Sets the item's respawn rate
+     * @param rate the rate in ticks or 0 if the default should be used
+     */
+    void setDespawnRate(int rate);
+
+    /**
+     * @return whether the item is affected by gravity
+     */
+    boolean isAffectedByGravity();
+
+    /**
+     * Sets whether the item should be affected by gravity
+     * @param gravity whether the item should be affected by gravity
+     */
+    void setAffectedByGravity(boolean gravity);
+
+    /**
+     * Prevents players from picking up this item.
+     */
+    default void noPickup() {
+        setPickupDelay(32767);
+    }
 }
-- 
2.26.2

