From 52cc5d5f46e1df202a20c739710feac24c32f1b4 Mon Sep 17 00:00:00 2001
From: RoccoDev <roccodev.business@gmail.com>
Date: Mon, 21 Sep 2020 16:24:06 +0200
Subject: [PATCH] Item Projectile effects


diff --git a/src/main/java/dev/rocco/kig/paper/api/particle/ItemProjectileEffect.java b/src/main/java/dev/rocco/kig/paper/api/particle/ItemProjectileEffect.java
index bdc5b525..47830fe5 100644
--- a/src/main/java/dev/rocco/kig/paper/api/particle/ItemProjectileEffect.java
+++ b/src/main/java/dev/rocco/kig/paper/api/particle/ItemProjectileEffect.java
@@ -9,14 +9,14 @@ public class ItemProjectileEffect extends ProjectileEffect {
     private ItemStack stack;
 
     public ItemProjectileEffect(ItemStack stack) {
-        super(1);
+        super(2);
         this.stack = stack;
     }
 
     @Override
     public void spawn(Location loc) {
         Item item = loc.getWorld().dropImmovableItem(loc, stack);
-        item.setDespawnRate(5);
+        item.setDespawnRate(3);
         item.noPickup();
         item.setAffectedByGravity(false);
         item.setVelocity(new Vector(0, 0, 0));
-- 
2.26.2

