From 3c51c28a868a7be8f4cbcdfa6638d9335ec86738 Mon Sep 17 00:00:00 2001
From: RoccoDev <roccodev.business@gmail.com>
Date: Mon, 21 Sep 2020 20:41:46 +0200
Subject: [PATCH] Nickname player info injection


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index d02fe560..3e340f79 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -1193,6 +1193,27 @@ public interface Player extends HumanEntity, Conversable, CommandSender, Offline
     @Deprecated
     public void sendTitle(String title, String subtitle);
 
+    /**
+     * Sets the player's nickname.
+     * @param nickname the nickname or {@code null} to reset it.
+     */
+    void setNickname(String nickname);
+
+    /**
+     * @return whether the player is nicknamed
+     */
+    boolean isNicknamed();
+
+    /**
+     * @return the player's nickname, or {@code null} if the player isn't nicknamed.
+     */
+    String getNickname();
+
+    /**
+     * @return the player's nickname if the player is nicknamed, or the player's real name
+     */
+    String getEffectiveName();
+
     /**
      * Resets the title displayed to the player.
      */
-- 
2.26.2

