From 54d4aee5511c6f25c96a983e7f8ddba93c14344e Mon Sep 17 00:00:00 2001
From: RoccoDev <roccodev.business@gmail.com>
Date: Tue, 22 Sep 2020 12:12:33 +0200
Subject: [PATCH] Player list entry self update


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 3e340f79..4ccd1e99 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -1214,6 +1214,13 @@ public interface Player extends HumanEntity, Conversable, CommandSender, Offline
      */
     String getEffectiveName();
 
+    /**
+     * Updates the user's player list entry.
+     * This change only applies to the current player.
+     * To change the display for other players, use {@link Player#hidePlayer(Player)} and {@link Player#showPlayer(Player)} instead.
+     */
+    void updateSelfPlayerInfo();
+
     /**
      * Resets the title displayed to the player.
      */
-- 
2.26.2

