From 3533ee468377034dec39c78caa9c73f9c0b62665 Mon Sep 17 00:00:00 2001
From: RoccoDev <roccodev.business@gmail.com>
Date: Sun, 27 Sep 2020 18:58:06 +0200
Subject: [PATCH] Custom Knockback API


diff --git a/src/main/java/dev/rocco/kig/paper/api/velocity/KnockbackValues.java b/src/main/java/dev/rocco/kig/paper/api/velocity/KnockbackValues.java
new file mode 100644
index 00000000..7293debf
--- /dev/null
+++ b/src/main/java/dev/rocco/kig/paper/api/velocity/KnockbackValues.java
@@ -0,0 +1,18 @@
+package dev.rocco.kig.paper.api.velocity;
+
+public interface KnockbackValues {
+    float getHorizontalMultiplier();
+    float getVerticalMultiplier();
+    float getFriction();
+    float getSprintingHorizontalMultiplier();
+    float getSprintingVerticalMultiplier();
+    void setHorizontalMultiplier(float multiplier);
+    void setVerticalMultiplier(float multiplier);
+    void setSprintingVerticalMultiplier(float multiplier);
+    void setSprintingHorizontalMultiplier(float multiplier);
+    void resetHorizontalMultiplier();
+    void resetVerticalMultiplier();
+    void resetFriction();
+    void resetSprintingHorizontalMultiplier();
+    void resetSprintingVerticalMultiplier();
+}
diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 57a8d048..c00fce8d 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -13,6 +13,7 @@ import java.util.UUID;
 import java.util.logging.Logger;
 
 import dev.rocco.kig.paper.api.skin.SkinCache;
+import dev.rocco.kig.paper.api.velocity.KnockbackValues;
 import org.bukkit.Warning.WarningState;
 import org.bukkit.command.CommandException;
 import org.bukkit.command.CommandMap;
@@ -1178,5 +1179,8 @@ public final class Bukkit {
     public static SkinCache getSkinCache() {
         return server.getSkinCache();
     }
+    public static KnockbackValues getKnockbackSettings() {
+        return server.getKnockbackSettings();
+    }
     // KigPaper end
 }
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 0f6383c1..5d24b212 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -13,6 +13,7 @@ import java.util.UUID;
 import java.util.logging.Logger;
 
 import dev.rocco.kig.paper.api.skin.SkinCache;
+import dev.rocco.kig.paper.api.velocity.KnockbackValues;
 import org.bukkit.Warning.WarningState;
 import org.bukkit.command.CommandException;
 import org.bukkit.command.CommandMap;
@@ -1023,5 +1024,9 @@ public interface Server extends PluginMessageRecipient {
      * @return async utilities for caching skins
      */
     SkinCache getSkinCache();
+    /**
+     * @return the server's custom knockback values
+     */
+    KnockbackValues getKnockbackSettings();
     // KigPaper end
 }
-- 
2.26.2

