From fecd5035365ea4f35b78fca562fd9636cfe83c8f Mon Sep 17 00:00:00 2001
From: RoccoDev <roccodev.business@gmail.com>
Date: Sun, 27 Sep 2020 19:51:16 +0200
Subject: [PATCH] Knockback friction API


diff --git a/src/main/java/dev/rocco/kig/paper/api/velocity/KnockbackValues.java b/src/main/java/dev/rocco/kig/paper/api/velocity/KnockbackValues.java
index 7293debf..a19ee197 100644
--- a/src/main/java/dev/rocco/kig/paper/api/velocity/KnockbackValues.java
+++ b/src/main/java/dev/rocco/kig/paper/api/velocity/KnockbackValues.java
@@ -10,6 +10,7 @@ public interface KnockbackValues {
     void setVerticalMultiplier(float multiplier);
     void setSprintingVerticalMultiplier(float multiplier);
     void setSprintingHorizontalMultiplier(float multiplier);
+    void setFriction(float friction);
     void resetHorizontalMultiplier();
     void resetVerticalMultiplier();
     void resetFriction();
-- 
2.26.2

