From 056b77c9dec1d4a2df3c1b774ddb35632ddea40f Mon Sep 17 00:00:00 2001
From: RoccoDev <hey@rocco.dev>
Date: Sun, 3 Jan 2021 14:06:47 +0100
Subject: [PATCH] Allow entities to be immune to stuck arrows


diff --git a/src/main/java/org/bukkit/entity/LivingEntity.java b/src/main/java/org/bukkit/entity/LivingEntity.java
index 48e25089..d93c9c93 100644
--- a/src/main/java/org/bukkit/entity/LivingEntity.java
+++ b/src/main/java/org/bukkit/entity/LivingEntity.java
@@ -392,4 +392,15 @@ public interface LivingEntity extends Entity, Damageable, ProjectileSource {
      * @return whether the operation was successful
      */
     public boolean setLeashHolder(Entity holder);
+
+    /**
+     * Sets whether the entity supports having arrows stuck to itself.
+     * @param capability whether arrows can stick to the entity
+     */
+    void setCanArrowsBeStuck(boolean capability);
+
+    /**
+     * Returns whether arrows can stick to the entity
+     */
+    boolean getCanArrowsBeStuck();
 }
-- 
2.28.0

