From c9ee2a6187479f6c5bc9dd5cb7fb4cd0dcc90fe3 Mon Sep 17 00:00:00 2001
From: RoccoDev <hey@rocco.dev>
Date: Sat, 16 Jan 2021 18:25:23 +0100
Subject: [PATCH] Add dismount cause to EntityDismountEvent


diff --git a/src/main/java/org/spigotmc/event/entity/EntityDismountEvent.java b/src/main/java/org/spigotmc/event/entity/EntityDismountEvent.java
index b35c7c18..8719db2b 100644
--- a/src/main/java/org/spigotmc/event/entity/EntityDismountEvent.java
+++ b/src/main/java/org/spigotmc/event/entity/EntityDismountEvent.java
@@ -18,11 +18,13 @@ public class EntityDismountEvent extends EntityEvent implements Cancellable // P
     private static final HandlerList handlers = new HandlerList();
     private boolean cancelled;
     private final Entity dismounted;
+    private final DismountCause cause;
 
-    public EntityDismountEvent(Entity what, Entity dismounted)
+    public EntityDismountEvent(Entity what, Entity dismounted, DismountCause cause) // KigPaper - add cause
     {
         super( what );
         this.dismounted = dismounted;
+        this.cause = cause;
     }
 
     public Entity getDismounted()
@@ -52,4 +54,15 @@ public class EntityDismountEvent extends EntityEvent implements Cancellable // P
         this.cancelled = cancelled;
     }
     // PaperSpigot end
+
+    // KigPaper start - distinguish between natural dismount and Bukkit API dismount
+    public enum DismountCause {
+        NATURAL,
+        BUKKIT
+    }
+
+    public DismountCause getCause() {
+        return cause;
+    }
+    // KigPaper end
 }
-- 
2.28.0

