From 4e55285322da0bffee4d2a7feb3d7c814aa47ca8 Mon Sep 17 00:00:00 2001
From: RoccoDev <hey@rocco.dev>
Date: Wed, 20 Jan 2021 17:28:08 +0100
Subject: [PATCH] Add World getTileEntities


diff --git a/src/main/java/org/bukkit/World.java b/src/main/java/org/bukkit/World.java
index a8137e88..3d7b2686 100644
--- a/src/main/java/org/bukkit/World.java
+++ b/src/main/java/org/bukkit/World.java
@@ -1,6 +1,8 @@
 package org.bukkit;
 
 import java.io.File;
+
+import org.bukkit.block.BlockState;
 import org.bukkit.generator.ChunkGenerator;
 import java.util.Collection;
 import java.util.HashMap;
@@ -1192,6 +1194,7 @@ public interface World extends PluginMessageRecipient, Metadatable {
      */
     public boolean isGameRule(String rule);
 
+    // KigPaper start
     /**
      * Drops an item at the specified {@link Location}, setting its velocity to 0.
      *
@@ -1201,6 +1204,12 @@ public interface World extends PluginMessageRecipient, Metadatable {
      */
     Item dropImmovableItem(Location location, ItemStack item);
 
+    /**
+     * Returns the list of tile entities currently residing in the world.
+     */
+    List<BlockState> getTileEntities();
+    // KigPaper end
+
     // Spigot start
     public class Spigot
     {
-- 
2.28.0

