From e4d669a36a786a69634f59d097a0915cc5ba9aa5 Mon Sep 17 00:00:00 2001
From: RoccoDev <hey@rocco.dev>
Date: Mon, 8 Feb 2021 16:49:18 +0100
Subject: [PATCH] Allow disabling cursors in CraftMapRenderer


diff --git a/src/main/java/org/bukkit/map/MapRenderer.java b/src/main/java/org/bukkit/map/MapRenderer.java
index 322d0ce3..442fe062 100644
--- a/src/main/java/org/bukkit/map/MapRenderer.java
+++ b/src/main/java/org/bukkit/map/MapRenderer.java
@@ -6,7 +6,7 @@ import org.bukkit.entity.Player;
  * Represents a renderer for a map.
  */
 public abstract class MapRenderer {
-
+    private boolean drawCursors = true; // KigPaper
     private boolean contextual;
 
     /**
@@ -53,4 +53,20 @@ public abstract class MapRenderer {
      */
     abstract public void render(MapView map, MapCanvas canvas, Player player);
 
+    // KigPaper start
+    /**
+     * Returns whether this renderer should draw cursors. Implementors should respect this choice.
+     */
+    public boolean getDrawCursors() {
+        return drawCursors;
+    }
+
+    /**
+     * Sets whether to draw cursors
+     * @param drawCursors whether this renderer should draw cursors
+     */
+    public void setDrawCursors(boolean drawCursors) {
+        this.drawCursors = drawCursors;
+    }
+    // KigPaper end
 }
-- 
2.28.0

