From c572a9adb1c286ef590e89a94df55f7e33477350 Mon Sep 17 00:00:00 2001
From: RoccoDev <hey@rocco.dev>
Date: Thu, 18 Feb 2021 20:37:19 +0100
Subject: [PATCH] ItemStack max stack size API


diff --git a/src/main/java/org/bukkit/inventory/ItemStack.java b/src/main/java/org/bukkit/inventory/ItemStack.java
index 957a0b56..a3c65d2a 100644
--- a/src/main/java/org/bukkit/inventory/ItemStack.java
+++ b/src/main/java/org/bukkit/inventory/ItemStack.java
@@ -280,7 +280,7 @@ public class ItemStack implements Cloneable, ConfigurationSerializable {
      */
     @Utility
     public int getMaxStackSize() {
-        if(this.customMaxStackSize != -1) return this.customMaxStackSize; // KigPaper
+        if(this.hasCustomMaxStackSize()) return this.customMaxStackSize; // KigPaper
         Material material = getType();
         if (material != null) {
             return material.getMaxStackSize();
-- 
2.28.0

