From cf80657122752620cba713da096ef066e9606f3d Mon Sep 17 00:00:00 2001
From: RoccoDev <hey@rocco.dev>
Date: Sat, 27 Feb 2021 13:29:38 +0100
Subject: [PATCH] Add Actionbar API


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 4ccd1e99..bb579294 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -2,6 +2,7 @@ package org.bukkit.entity;
 
 import java.net.InetSocketAddress;
 
+import net.md_5.bungee.api.chat.BaseComponent;
 import org.bukkit.Achievement;
 import org.bukkit.ChatColor;
 import org.bukkit.Effect;
@@ -1193,6 +1194,7 @@ public interface Player extends HumanEntity, Conversable, CommandSender, Offline
     @Deprecated
     public void sendTitle(String title, String subtitle);
 
+    // KigPaper start
     /**
      * Sets the player's nickname.
      * @param nickname the nickname or {@code null} to reset it.
@@ -1214,6 +1216,7 @@ public interface Player extends HumanEntity, Conversable, CommandSender, Offline
      */
     String getEffectiveName();
 
+
     /**
      * Updates the user's player list entry.
      * This change only applies to the current player.
@@ -1221,6 +1224,13 @@ public interface Player extends HumanEntity, Conversable, CommandSender, Offline
      */
     void updateSelfPlayerInfo();
 
+    /**
+     * Sends a message to the player's actionbar.
+     * @param message the message to send
+     */
+    void sendActionbarMessage(BaseComponent... message);
+    // KigPaper end
+
     /**
      * Resets the title displayed to the player.
      */
-- 
2.28.0

