From 7257125a1d68271c909a9eac66d469b7f83a7756 Mon Sep 17 00:00:00 2001
From: RoccoDev <hey@rocco.dev>
Date: Wed, 10 Mar 2021 14:17:12 +0100
Subject: [PATCH] Add World getEntityById API


diff --git a/src/main/java/org/bukkit/World.java b/src/main/java/org/bukkit/World.java
index 9edac15d..68f5d488 100644
--- a/src/main/java/org/bukkit/World.java
+++ b/src/main/java/org/bukkit/World.java
@@ -1218,6 +1218,13 @@ public interface World extends PluginMessageRecipient, Metadatable {
      * Returns the list of tile entities currently residing in the world.
      */
     List<BlockState> getTileEntities();
+
+    /**
+     * Returns an entity by its internal ID
+     * @param id the entity's internal ID, as returned by {@link Entity#getEntityId()}.
+     * @return the entity, or {@code null} if it could not be found
+     */
+    Entity getEntityById(int id);
     // KigPaper end
 
     // Spigot start
-- 
2.28.0

