From ae4359a6bf278ecb7d404c35b0987acd92bd26ed Mon Sep 17 00:00:00 2001
From: RoccoDev <hey@rocco.dev>
Date: Wed, 10 Mar 2021 20:21:55 +0100
Subject: [PATCH] Add LivingEntity clear last attacked API


diff --git a/src/main/java/org/bukkit/entity/LivingEntity.java b/src/main/java/org/bukkit/entity/LivingEntity.java
index d93c9c93..cc8d81ca 100644
--- a/src/main/java/org/bukkit/entity/LivingEntity.java
+++ b/src/main/java/org/bukkit/entity/LivingEntity.java
@@ -393,6 +393,7 @@ public interface LivingEntity extends Entity, Damageable, ProjectileSource {
      */
     public boolean setLeashHolder(Entity holder);
 
+    // KigPaper start
     /**
      * Sets whether the entity supports having arrows stuck to itself.
      * @param capability whether arrows can stick to the entity
@@ -403,4 +404,10 @@ public interface LivingEntity extends Entity, Damageable, ProjectileSource {
      * Returns whether arrows can stick to the entity
      */
     boolean getCanArrowsBeStuck();
+
+    /**
+     * Clears the last attacked player for this entity. Used to reset pathfinder goals for tameables owned by this entity.
+     */
+    void clearLastAttacked();
+    // KigPaper end
 }
-- 
2.28.0

