From 4df794f6805c116c0d6b0c41d287d0a561441112 Mon Sep 17 00:00:00 2001
From: RoccoDev <hey@rocco.dev>
Date: Thu, 18 Mar 2021 19:55:31 +0100
Subject: [PATCH] Add yaw and pitch to World spawn location


diff --git a/src/main/java/org/bukkit/World.java b/src/main/java/org/bukkit/World.java
index 68f5d488..9a23d4cb 100644
--- a/src/main/java/org/bukkit/World.java
+++ b/src/main/java/org/bukkit/World.java
@@ -496,15 +496,31 @@ public interface World extends PluginMessageRecipient, Metadatable {
      */
     public Location getSpawnLocation();
 
+    // KigPaper start
+    /**
+     * Sets the spawn location of the world
+     *
+     * @param x X coordinate
+     * @param y Y coordinate
+     * @param z Z coordinate
+     * @return True if it was successfully set.
+     */
+    default boolean setSpawnLocation(int x, int y, int z) {
+        return setSpawnLocation(x, y, z, 0, 0);
+    }
+
     /**
      * Sets the spawn location of the world
      *
      * @param x X coordinate
      * @param y Y coordinate
      * @param z Z coordinate
+     * @param yaw yaw rotation
+     * @param pitch pitch rotation
      * @return True if it was successfully set.
      */
-    public boolean setSpawnLocation(int x, int y, int z);
+    public boolean setSpawnLocation(int x, int y, int z, float yaw, float pitch);
+    // KigPaper end
 
     /**
      * Gets the relative in-game time of this world.
diff --git a/src/main/java/org/bukkit/command/defaults/SetWorldSpawnCommand.java b/src/main/java/org/bukkit/command/defaults/SetWorldSpawnCommand.java
index 8bec19c9..dde33815 100644
--- a/src/main/java/org/bukkit/command/defaults/SetWorldSpawnCommand.java
+++ b/src/main/java/org/bukkit/command/defaults/SetWorldSpawnCommand.java
@@ -62,7 +62,7 @@ public class SetWorldSpawnCommand extends VanillaCommand {
             return false;
         }
 
-        world.setSpawnLocation(x, y, z);
+        world.setSpawnLocation(x, y, z, player == null ? 0f : player.getLocation().getYaw(), player == null ? 0f : player.getLocation().getPitch()); // KigPaper - add yaw/pitch
 
         Command.broadcastCommandMessage(sender, "Set world " + world.getName() + "'s spawnpoint to (" + x + ", " + y + ", " + z + ")");
         return true;
-- 
2.28.0

