From 23d997baa9d07c54dd503a895805dc6f71e34c46 Mon Sep 17 00:00:00 2001
From: RoccoDev <hey@rocco.dev>
Date: Sun, 4 Apr 2021 18:15:11 +0200
Subject: [PATCH] Add particle emitter API to World.Spigot#playEffect


diff --git a/src/main/java/dev/rocco/kig/paper/api/particle/ParticleProjectileEffect.java b/src/main/java/dev/rocco/kig/paper/api/particle/ParticleProjectileEffect.java
index df49de62..53f01b0b 100644
--- a/src/main/java/dev/rocco/kig/paper/api/particle/ParticleProjectileEffect.java
+++ b/src/main/java/dev/rocco/kig/paper/api/particle/ParticleProjectileEffect.java
@@ -2,6 +2,7 @@ package dev.rocco.kig.paper.api.particle;
 
 import org.bukkit.Effect;
 import org.bukkit.Location;
+import org.bukkit.entity.Player;
 
 public class ParticleProjectileEffect extends ProjectileEffect {
     private Effect effect;
@@ -10,7 +11,11 @@ public class ParticleProjectileEffect extends ProjectileEffect {
 
     @Override
     public void spawn(Location loc) {
-        loc.getWorld().spigot().playEffect(loc, effect, id, data, offsetX, offsetY, offsetZ, speed, count, radius);
+        spawn(loc, null);
+    }
+
+    public void spawn(Location loc, Player emitter) {
+        loc.getWorld().spigot().playEffect(loc, effect, id, data, offsetX, offsetY, offsetZ, speed, count, radius, emitter);
     }
 
     private ParticleProjectileEffect() {
diff --git a/src/main/java/org/bukkit/World.java b/src/main/java/org/bukkit/World.java
index 9a23d4cb..ef5399da 100644
--- a/src/main/java/org/bukkit/World.java
+++ b/src/main/java/org/bukkit/World.java
@@ -1286,8 +1286,14 @@ public interface World extends PluginMessageRecipient, Metadatable {
          */
         public void playEffect(Location location, Effect effect, int id, int data, float offsetX, float offsetY, float offsetZ, float speed, int particleCount, int radius)
         {
-            throw new UnsupportedOperationException( "Not supported yet." );
+            playEffect(location, effect, id, data, offsetX, offsetY, offsetZ, speed, particleCount, radius, null);
+        }
+
+        // KigPaper start
+        public void playEffect(Location location, Effect effect, int id, int data, float offsetX, float offsetY, float offsetZ, float speed, int particleCount, int radius, Player emitter) {
+            throw new UnsupportedOperationException("Not supported yet.");
         }
+        // KigPaper end
         
         /**
          * Strikes lightning at the given {@link Location} and possibly without sound
-- 
2.28.0

