From 57ddb45602aef18d276f2bb4c316a3ec7c2d7d50 Mon Sep 17 00:00:00 2001
From: RoccoDev <hey@rocco.dev>
Date: Wed, 14 Apr 2021 14:00:10 +0200
Subject: [PATCH] Add exit code to shutdown API


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 9c4650a9..1f4c1da1 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -1185,5 +1185,8 @@ public final class Bukkit {
     public static void setMaxPlayers(int count) {
         server.setMaxPlayers(count);
     }
+    public static void shutdown(int exitCode) {
+        server.shutdown(exitCode);
+    }
     // KigPaper end
 }
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 375d41ce..554e07f8 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -597,6 +597,15 @@ public interface Server extends PluginMessageRecipient {
      */
     public void shutdown();
 
+    // KigPaper start
+
+    /**
+     * Shuts the server down, stopping everything.
+     * @param exitCode the exit code to use
+     */
+    void shutdown(int exitCode);
+    // KigPaper end
+
     /**
      * Broadcasts the specified message to every user with the given
      * permission name.
-- 
2.28.0

