From 6e8baa0bbcd7be761464687973284c83aae1dd6c Mon Sep 17 00:00:00 2001
From: RoccoDev <hey@rocco.dev>
Date: Sun, 23 May 2021 00:25:07 +0200
Subject: [PATCH] Add Player last activity API


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index c1768637..f2a07235 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -1237,6 +1237,11 @@ public interface Player extends HumanEntity, Conversable, CommandSender, Offline
      * Essentially this returns a spoofed UUID if the player is nicked, or the user's normal UUID otherwise.
      */
     UUID getUniqueIdForDisplay();
+
+    /**
+     * Returns the player's last activity timestamp, according to the "idle timeout" checks.
+     */
+    long getLastActivity();
     // KigPaper end
 
     /**
-- 
2.31.1

