From 64c0d7a7c41f262da4d30d5f0ea347b13a3d0530 Mon Sep 17 00:00:00 2001
From: RoccoDev <hey@rocco.dev>
Date: Wed, 26 May 2021 23:59:24 +0200
Subject: [PATCH] Add DisconnectReason to PlayerKickEvent


diff --git a/src/main/java/dev/rocco/kig/paper/api/event/DisconnectReason.java b/src/main/java/dev/rocco/kig/paper/api/event/DisconnectReason.java
new file mode 100644
index 00000000..62973ba2
--- /dev/null
+++ b/src/main/java/dev/rocco/kig/paper/api/event/DisconnectReason.java
@@ -0,0 +1,67 @@
+package dev.rocco.kig.paper.api.event;
+
+/**
+ * Reason for {@link org.bukkit.event.player.PlayerQuitEvent} and {@link org.bukkit.event.player.PlayerKickEvent}.
+ */
+public enum DisconnectReason {
+    // Client-bound
+
+    /**
+     * The client disconnected normally
+     */
+    DISCONNECTED,
+    /**
+     * Read timed out on the connection
+     */
+    TIMED_OUT,
+    /**
+     * An unknown exception has occurred in the connection
+     */
+    CONNECTION_EXCEPTION,
+
+    // Kicks
+
+    /**
+     * The server is shutting down
+     */
+    SERVER_SHUTTING_DOWN,
+    /**
+     * The player has died and the server is in hardcore mode
+     */
+    HARDCORE_DEATH,
+    /**
+     * The player was kicked because the same player logged in from a different location
+     */
+    KICK_ALREADY_LOGGED_IN,
+    /**
+     * The player sent an invalid packet (unknown, malformed etc.)
+     */
+    KICK_BAD_PACKET,
+    /**
+     * The player performed certain actions too quickly. Doesn't include spamming
+     */
+    KICK_RATE_LIMIT,
+    /**
+     * The player sent an invalid packet, and this could have caused harm to the server
+     */
+    KICK_BAD_PACKET_SEVERE,
+    /**
+     * The player has been idle for too long
+     */
+    KICK_IDLE,
+    /**
+     * The player has been sending messages/commands too quickly
+     */
+    KICK_SPAM,
+    /**
+     * The player was hovering for too long
+     */
+    KICK_FLYING,
+
+    // Custom
+
+    /**
+     * The player was kicked by a plugin
+     */
+    KICK_PLUGIN
+}
diff --git a/src/main/java/org/bukkit/event/player/PlayerKickEvent.java b/src/main/java/org/bukkit/event/player/PlayerKickEvent.java
index 39e81b67..a3185e77 100644
--- a/src/main/java/org/bukkit/event/player/PlayerKickEvent.java
+++ b/src/main/java/org/bukkit/event/player/PlayerKickEvent.java
@@ -1,5 +1,6 @@
 package org.bukkit.event.player;
 
+import dev.rocco.kig.paper.api.event.DisconnectReason;
 import org.bukkit.entity.Player;
 import org.bukkit.event.Cancellable;
 import org.bukkit.event.HandlerList;
@@ -10,14 +11,16 @@ import org.bukkit.event.HandlerList;
 public class PlayerKickEvent extends PlayerEvent implements Cancellable {
     private static final HandlerList handlers = new HandlerList();
     private String leaveMessage;
+    private final DisconnectReason disconnectReason; // KigPaper
     private String kickReason;
     private Boolean cancel;
 
-    public PlayerKickEvent(final Player playerKicked, final String kickReason, final String leaveMessage) {
+    public PlayerKickEvent(final Player playerKicked, final String kickReason, final String leaveMessage, final DisconnectReason disconnectReason) {
         super(playerKicked);
         this.kickReason = kickReason;
         this.leaveMessage = leaveMessage;
         this.cancel = false;
+        this.disconnectReason = disconnectReason;
     }
 
     /**
@@ -64,6 +67,15 @@ public class PlayerKickEvent extends PlayerEvent implements Cancellable {
         this.leaveMessage = leaveMessage;
     }
 
+    // KigPaper start
+    /**
+     * Returns the machine-friendly disconnect reason
+     */
+    public DisconnectReason getDisconnectReason() {
+        return disconnectReason;
+    }
+    // KigPaper end
+
     @Override
     public HandlerList getHandlers() {
         return handlers;
-- 
2.31.1

