From d4306b8f2563018a75f9cff2ca107d85737e3927 Mon Sep 17 00:00:00 2001
From: RoccoDev <hey@rocco.dev>
Date: Tue, 1 Jun 2021 13:55:14 +0200
Subject: [PATCH] Player DisconnectReason kick API


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index f2a07235..d3db3cc4 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -3,6 +3,7 @@ package org.bukkit.entity;
 import java.net.InetSocketAddress;
 import java.util.UUID;
 
+import dev.rocco.kig.paper.api.event.DisconnectReason;
 import net.md_5.bungee.api.chat.BaseComponent;
 import org.bukkit.Achievement;
 import org.bukkit.ChatColor;
@@ -1242,6 +1243,14 @@ public interface Player extends HumanEntity, Conversable, CommandSender, Offline
      * Returns the player's last activity timestamp, according to the "idle timeout" checks.
      */
     long getLastActivity();
+
+    /**
+     * Kicks player with custom kick message and reason.
+     *
+     * @param reason machine-friendly kick reason
+     * @param message kick message
+     */
+    void kickPlayer(DisconnectReason reason, String message);
     // KigPaper end
 
     /**
-- 
2.31.1

