From 0955cce84a6f35616275b65a5ce9b2ad3aebb3d7 Mon Sep 17 00:00:00 2001
From: RoccoDev <hey@rocco.dev>
Date: Fri, 11 Jun 2021 23:02:36 +0200
Subject: [PATCH] Player register plugin channel API


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index d3db3cc4..15783045 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -20,6 +20,7 @@ import org.bukkit.WeatherType;
 import org.bukkit.command.CommandSender;
 import org.bukkit.conversations.Conversable;
 import org.bukkit.map.MapView;
+import org.bukkit.plugin.Plugin;
 import org.bukkit.plugin.messaging.PluginMessageRecipient;
 import org.bukkit.scoreboard.Scoreboard;
 // PaperSpigot start
@@ -1251,6 +1252,13 @@ public interface Player extends HumanEntity, Conversable, CommandSender, Offline
      * @param message kick message
      */
     void kickPlayer(DisconnectReason reason, String message);
+
+    /**
+     * Manually registers a plugin channel for this player (so that {@link Player#sendPluginMessage(Plugin, String, byte[])} will work on that channel)
+     * An exception is thrown if the player has 128 registered channels or more.
+     * @param channel the channel to register
+     */
+    void registerPluginChannel(String channel);
     // KigPaper end
 
     /**
-- 
2.31.1

