From dfdd1fff31cda3a7ffa3601ea3ac1c88a1631c6b Mon Sep 17 00:00:00 2001
From: RoccoDev <hey@rocco.dev>
Date: Thu, 9 Sep 2021 20:49:42 +0200
Subject: [PATCH] Add PlayerInteractUpdateEvent


diff --git a/src/main/java/dev/rocco/kig/paper/api/event/PlayerInteractUpdateEvent.java b/src/main/java/dev/rocco/kig/paper/api/event/PlayerInteractUpdateEvent.java
new file mode 100644
index 00000000..5bb1e571
--- /dev/null
+++ b/src/main/java/dev/rocco/kig/paper/api/event/PlayerInteractUpdateEvent.java
@@ -0,0 +1,46 @@
+package dev.rocco.kig.paper.api.event;
+
+import org.bukkit.block.Block;
+import org.bukkit.entity.Player;
+import org.bukkit.event.Cancellable;
+import org.bukkit.event.HandlerList;
+import org.bukkit.event.player.PlayerEvent;
+
+/**
+ * Fired when a block change is sent to the client as a result of player interaction.
+ */
+public class PlayerInteractUpdateEvent extends PlayerEvent implements Cancellable {
+    private static final HandlerList handlers = new HandlerList();
+    private boolean cancel;
+    private final int x, y, z;
+
+    public PlayerInteractUpdateEvent(Player who, int x, int y, int z) {
+        super(who);
+        this.x = x;
+        this.y = y;
+        this.z = z;
+    }
+
+    @Override
+    public boolean isCancelled() {
+        return cancel;
+    }
+
+    @Override
+    public void setCancelled(boolean cancel) {
+        this.cancel = cancel;
+    }
+
+    public Block getBlock() {
+        return player.getWorld().getBlockAt(x, y, z);
+    }
+
+    @Override
+    public HandlerList getHandlers() {
+        return handlers;
+    }
+
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+}
-- 
2.31.1

